Span structure for messaging scenarios (#220)
open-telemetry/oteps: text/trace/0220-messaging-semantic-conventions-span-structure.md

Reviewer: I don't see an "Open Questions" section like we have in our template: https://github.com/open-telemetry/oteps/blob/main/0000-template.md#open-questions

What is the status of the "adding links after span start" topic? Does the current version of this OTEP require that feature? If so how to specify that should be added as an open question.

## Motivation

Tracking the path of an individual message through a distributed system poses
several challenges. Messaging systems allow for asynchronous workflows, which
means that the stages of producing and consuming a message can be separated by
a considerable time gap (this can be minutes, hours, or days). Furthermore, one
cannot rely on consistent instrumentation across all parts of the system that
touch a message. Correlating producer and consumer stages are expected even when
the intermediary forwarding the message between them is not instrumented.
Finally, batching of messages can happen in many different parts of a message
processing workflow, be it batch publishing, batch receiving, batch processing,
or batch settling.

Despite all those challenges, requirements for instrumentation of messaging
scenarios are high. Besides correlating spans that model the different
processing stages of a message, it should also be possible to determine the
end-to-end latency of processing a message. If intermediaries are not
instrumented, this shouldn't impact the correlation of producer and consumer
stages. If, on the other hand, intermediaries are instrumented, spans from

> intermediary instrumentation should seamlessly integrate with producer and

Author: I added those clarifications.

consumer instrumentation.

This OTEP aims at proposing consistent guidelines for creating spans that model
the stages of the messaging flow, and correlating those in a way so that the
requirements sketched above can be met in a consistent way across messaging
scenarios and different messaging systems.

## Terminology

For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).

## Stages of producing and consuming messages

Producing and consuming a message involves five stages:

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  CR[Create] --> PU[Publish]
  end
  subgraph INTERMEDIARY
  direction TB
  end
  subgraph CONSUMER
  direction TB
  RE[Receive] --> PR[Process]
  PR --> SE[Settle]
  end
  PU --> INTERMEDIARY
  INTERMEDIARY --> RE

>   SE -..-> INTERMEDIARY

Reviewer 2: Something that I raised in https://github.com/open-telemetry/oteps/pull/192 as well. We should have settlement on the producer side as well imho.

Author: Thanks for pointing this out. Could you give some use cases for producer settlement, or links to messaging systems that support it?

Reviewer 2: AFAIK mostly "all" messaging systems have such a support? It's about getting an ACK that the message was received by the consumer (if peer-to-peer, but let's take this out of scope) or intermediary (broker) in order to allow different QoS (fire and forget/at most once, at least once, "exactly once").
For example, with AMQP 1.0 the sender sends the `transfer` frame, it can set `settled=true` flag already (so fire and forget) or `settled=false` and waiting for hearing back from the broker by receiving a `disposition` frame which settles the delivery (on the broker itself). So for example a messaging system like Apache Artemis has this.

Even the JMS specification has such a thing at API level (both with sync and asyn sends). See [async send](https://jakarta.ee/specifications/messaging/3.0/jakarta-messaging-spec-3.0.html#asynchronous-send) and more specifically this sentence.

> "In some Jakarta Messaging providers, a normal synchronous send involves sending the message to a remote Jakarta Messaging server and then waiting for an acknowledgement to be received before returning. It is expected that such a provider would implement an asynchronous send by sending the message to the remote Jakarta Messaging server and then returning without waiting for an acknowledgement. When the acknowledgement is received, the Jakarta Messaging provider would notify the application by invoking the onCompletion method on the application-specified CompletionListener object. If for some reason the acknowledgement is not received the Jakarta Messaging provider would notify the application by invoking the CompletionListener’s onException method."

Of course Kafka, the producer can set different levels of acks (not getting ack, getting ack from leader, getting ack when message is received from leader and replicated to followers). See official doc [here](https://kafka.apache.org/documentation/#producerconfigs_acks).

Author: Thanks for giving the links. We discussed this in today's workgroup meeting. To summarize, "producer settlement", according to our understanding, involves receiving an ACK on the producer side from the intermediary, confirming that the intermediary successfully received the message.

We identified two main scenarios, for modelling both we intend to look at it more from the user perspective than from the underlying messaging system perspective (like AMQP details, which will be covered by transport layer instrumentation).
1. For synchronous settlement, the user will initiate a publish operation, this operation will block until an ACK for each message is received. This should be covered by a single "Publish" span.
2. For asynchronous settlement, the user will initiate a publish operation, which returns before awaiting an ACK from the intermediary. The instrumentation in this case should reflect the way that status is returned to the user. If a callback is invoked, this could involve a separate span for this callback which is parented/linked to the initial "Publish" span. However, we think the possibilities here are very varied and bound to implementations of messaging SDKS, so we don't want to come up with generic conventions for those cases at this point.

Reviewer 2: > We discussed this in today's workgroup meeting

And I had conflicts as usual on my agenda. Sorry but there is a weekly meeting I have that time which is never cancelled :-(

Regarding the point 1 and 2 ...
How are you going to differentiate scenario 1 where sync message was sent, you have a "Publish" span only because an ACK was returned compared to scenario 2 where you have async message set, you have a "Publish" span  but the user asked for "fire and forget" so no ACK was got by the SDK. Won't they overlap? 
Instead, when a callback is called are you referring to create a "Settle" span linked to the "Publish"?

Author: This was discussed in today's messaging workgroup. The consensus in the group was, that also asynchronous settlement operations on the producer side should be modelled with a single span whenever possible, this span should end when the ACK was received.

In cases where this is not possible, a using a separate "settle" span makes sense. However, as this seems to be an edge case, we decided not to add additional complexity by specifying this but leave more flexibility to instrumentors.

attaching independent contexts with messages.

If a "Create" span exists for a message, its context must be attached to the
message. If no "Create" span exists for a message, the context of the related
"Publish" span must be attached to the message.

### Consumer

For many use cases, it is not possible to rely on the presence of "Process"
spans for correlating producer with consumer traces: there are cases where a
dedicated processing operation cannot be identified, or where processing
happens in a different trace. Furthermore, processing operations often are not
covered by messaging libraries and SDKs, but take place in application code.
Consistently creating spans for "Processing" operations would require either
effort from the application owner to correctly instrument those operations, or
additional capabilities of messaging libraries and SDKs (e. g. hooks for
processing callbacks, which can then be instrumented by the libraries or SDKs).

While it is possible to create "Process" spans and correlate those with
consumer traces in certain cases, this is not something that can be generally
required. Therefore, it is more feasible to require the creation of "Deliver"

> spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to

Reviewer 2: Shouldn't it be clarified in the above graph as well?

> spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to

Author: We don't give any directions about "Process" spans here, that's what this paragraph tries to point out. So, we also don't specify how those correlate with "Deliver" and "Receive" spans, and what's their parent context.

Probably we should have another discussion on one of the next Thursdays about "Process" spans.

> spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to

Author: Closing this, as I outlined reasons for not specifying process spans in another comment: https://github.com/open-telemetry/oteps/pull/220#discussion_r993981013

correlate producer with consumer traces.

#### Instrumenting push-based scenarios

In push-based consumer scenarios, the delivery of messages is not initiated by
the application code. Instead, callbacks or handlers are registered and then
called by messaging SDKs to forward messages to the application.

A "Deliver" span covers the call of such a callback or handler and should link
to the "Create" spans of all messages that are forwarded via the respective
call. Depending on the use case, "Deliver" spans can correlate with "Process"
spans or other spans modelling processing operations.

#### Instrumenting pull-based scenarios

In pull-based consumer scenarios, the delivery of messages is requested by the
application code. This usually involves a blocking call, which returns zero or
more messages on completion.

A "Receive" span covers such calls and should link to the "Create" spans of all
messages that are forwarded via the respective call. Depending on the use case,

> "Receive" spans can correlate with "Process" spans or other spans modelling

Reviewer 3: I think correlate is too ambiguous and we should avoid it whenever possible. Should it be either parent, link or both?

Author: We are ambiguous on purpose here. We don't specify any conventions for "Process" spans in this document (yet). Users might create those spans, but we leave it open whether they are created at all and how they are "correlated" to other spans.

Reviewer 3: It will save time to a lot of people if we take a decision on this regard.

Author: We refrained from adding conventions for "Process" spans for several reasons:
* It's hard to define what "processing" a message entails. It might be a single operation, it also might be multiple operations, sometimes messages might be received but not be processed at all. As there are many ways of how to process a message, we don't want to limit users by any strong requirements.
* One aim of the conventions is, that they can be implemented by messaging SDKs, without the need for users to add additional instrumentation. Once we require the creation of process spans, this requirement will be very hard (or in many cases even impossible) to satisfy by SDK instrumentation.
* Given the above considerations, it is definitely possible to add recommendations about creating process spans now or at a future point in time, maybe in v2 of these conventions. As there are many problems to tackle, we thought it's preferable to get a smaller set of conventions stable earlier instead of taking longer for a comprehensive set of conventions.

Reviewer 4: While I understand that a `Process` span is difficult to produce at an instrumentation level, I think it should still be defined if only to be marked as optional.  If not, then when they are added it will be done inconsistently.

Author: This was discussed in today's messaging workgroup. There's interest and value in adding recommendations about "Process" spans, however, for this OTEP we'd like to focus on the _minimum_ of spans and correlations that is required and backends can rely on and that can be provided by auto-instrumentation (at least one of "create" or "publish" on producer side, and at least one of "deliver" or "receive" on the consumer side). We'd like to keep this OTEP at this scope, as it already has a length that makes it challenging to review.

@tylerbenson Do you think that what's proposed in this OTEP can serve as a basis for adding recommendations about "process" spans? Would you be willing to come up with a proposal for adding "process" spans? Either based on this OTEP, or if that's not possible, highlight the conflicts you see?

Reviewer 4: I agree that the process span definition should be done independently as this effort is already sufficiently large and complex. I am putting more focus on the FAAS/Java SIGs at the moment so I don't have cycles to contribute a proposal for the process span, but happy to review when one is ready.

the settlement operation was manually triggered by the user or automatically
triggered by SDKs. SDKs will in some cases auto-settle messages in
push-scenarios when messages are delivered via callbacks. In this case, it is
recommended to create a parent span, so that a "Settle" span will be a sibling
of the related "Deliver" span.

Alternatively, an event can be created instead of a "Settle" span. Events could
be added to "Deliver" spans or to ambient spans.

"Settle" spans may link to "Create" spans of the messages that are settled,
however, for some settlement scenarios this is not feasible or possible.

## Proposed changes and additions to the messaging semantic conventions

### Operation name

The following operations related to messages are covered by these semantic
conventions:

| Operation name | Description |
|----------------|-------------|
| `publish`      | One or more messages are provided for publishing to an intermediary. |
| `create`       | A message is created. |
| `receive`      | One or more messages are requested by a consumer. |
| `deliver`      | One or more messages are passed to a consumer. |

> | `settle`       | One or more messages are settled. |

Reviewer 3: Should we suggest composite names like?

create_publish
receive_process
...

Author: What benefit would one get from it? It's not clear to me.

Reviewer 3: because there are many cases where a message is created and published inline... Unless we point to Create being used by app code and publish to the lib actually managing the connection. 
If this is the case, it would be a nice clarification.

Author: For the "create_publish" combination, this is already covered by the supported case of a "publish" span that is of kind "PRODUCER". In this case we know that no "create" spans are present.

My stance is, that we shouldn't suggest composite names for other cases. Firstly, because a set of five distinct operations is much easier to grasp than a set of operations enlarged by composites. If there's "receive_process", there's probably a "deliver_process", and the people might ask for "receive_process_settle". Secondly, if you collapse operations, like e. g. "receive_process", you lose the possibility to model cases where the "receive" operation succeeds, but the "process" operation fails.

However, I'm open towards other opinions.

Author: This was discussed in today's messaging workgroup. There are cases where composite names would make sense, however, we're not sure if there are enough use-cases to merit the additional complexity.

We decided to stick with simple operation names for now, as composite names can be added on top of that in the future if there's need for it (whereas it's much harder the harder way around).

#### Consumer spans

##### Push-based scenarios

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application
code.

##### Pull-based scenarios

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code.

##### General considerations

"Deliver" or "Receive" spans MUST NOT be created for messages which are not
forwarded to the application, but are pre-fetched or cached by messaging
libraries or SDKs.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received).  For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
the "Create" span for the message.

#### Settlement spans

"Settle" spans or events SHOULD be created for every manually or automatically
triggered settlement operation. A single "Settle" span can account for a
single message or for multiple messages (in case messages are passed for

> settling as batches). For each message it accounts for, the "Settle" span

Reviewer 2: actually you could even got more messages one by one (single message) and then using a settlement for ALL of them. So it's not a batch but you are anyway using one settlement for more messages. Does the above sentence take this into account?

Author: I think you're referring to checkpoint-based settlement. 

For this case, I think we can't really expect to link to all messages that are settled by settling one message. The settlement span may have a sequence number attribute though, which indicates to what point the checkpoint was set.

Author: That's from the _Settlement_ section above:
> "Settle" spans may link to "Create" or "Publish" spans of the messages that are settled, however, for some settlement scenarios this is not feasible or possible.

In the end, this is a best-effort recommendation. Ideally, settlement spans have links to all the messages that are settled by the operation. However, in a considerable number of cases (and for checkpoint-based settlement in general) this is not feasible. That's why this is only put forth as a MAY, which is the weakest requirement level.

Reviewer 2: Ok makes sense. Maybe it could be anyway useful to improve the [checkpoint-based settlement](https://github.com/open-telemetry/oteps/blob/main/text/trace/0173-messaging-semantic-conventions.md#checkpoint-based-settlement) section because something like that can happen with AMQP 1.0 (not just Kafka). The `disposition` frame allows you to specify `first` and `last` fields with delivery IDs you are settling. It means even more complex settlement than the checkpoint, because it could settle sparse range of messages.

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  PM1[Publish m1]
  PM2[Publish m2]
  end
  subgraph CONSUMER
  direction TB
  A[Ambient]-->R[Receive]
  A-->SM1[Settle m1]
  A-->SM2[Settle m2]
  end
  PM1-. link .->R;
  PM2-. link .->R;
```

## Future possibilities

### Intermediary instrumentation

While intermediary instrumentation is not directly covered by the conventions
in this document, it certainly is necessary to keep the proposed conventions
extensible so that intermediary instrumentation can be easily added in a way
that integrates well with producer and consumer instrumentation.

The diagram below gives an example of how intermediary instrumentation can be
added. The fact that producers and consumers are consistently correlated by
links across all scenarios provides maximal flexibility for adding intermediary
instrumentation.

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  PM1[Publish m1]
  PM2[Publish m2]
  end
  subgraph INTERMEDIARY
  direction TB
  EQ[Enqueue]
  DQ[Dequeue]
  end
  subgraph CONSUMER
  direction TB
  D[Deliver]-->PRM1[Process m1]
  D-->PRM2[Process m2]
  end
  PM1-. link .->D;
  PM2-. link .->D;
  PM1-->EQ;
  PM2-->EQ;

>   DQ-->D;

Reviewer 2: Not sure this is a clear picture explaining how things work at messaging level. It depends on what we mean as "enqueue" and "dequeue" because when you have an intermediary in the middle, in the end you have a contract between producer and intermediary, and one between intermediary and consumer. 
So the intermediary "receive"(s) the message from producer (so you could see it as a consumer) and then it "publish"(es) the message to the consumer (so you could see it as a producer).

Author: I can change those names. I didn't want to use "Publish" and "Receive", as those are names we use for producer and consumer spans, so I wanted to avoid confusion.

Maybe I should just have empty boxes in the intermediary to avoid any confusion?

Reviewer 2: I would just remove those boxes and not having them empty. Just having a "gray" intermediary box without the need to highlight what's in there (even because tracing within intermediary is out of scope for now).

Author: I'd like to leave the example here, because it shows that the proposed the solution can easily extended to accomodate intermediary instrumentation.

The boxes inside the intermediary just mean to highlight that there exist some (yet undefined) span in the intermediary that acts as parent/child for the consumer and producer spans.

However, I'm open to further adjust the diagram if it makes things clearer.

Reviewer 2: > I'd like to leave the example here, because it shows that the proposed the solution can easily extended to accomodate intermediary instrumentation.

I am totally fine with leaving the example.

> The boxes inside the intermediary just mean to highlight that there exist some (yet undefined) span in the intermediary that acts as parent/child for the consumer and producer spans.

Tbh the empty boxes can just confuse people because they don't add/clarify nothing.

  PR --> SE[Settle]
  end
  PU --> INTERMEDIARY
  INTERMEDIARY --> RE
  SE -..-> INTERMEDIARY
```

1. The producer creates a message.
2. The producer publishes the message to an intermediary.
3. The consumer receives the message from an intermediary.
4. The consumer processes the message.
5. The consumer settles the message by notifying the intermediary that the
   message was processed. In some cases (fire-and-forget), the settlement stage
   does not exist.

The semantic conventions described below define how to model those stages with
spans.

## Trace structure

### Producer

Producers are responsible for attaching a creation context to a message.
Subsequent consumers will use this context to link consumer traces to producer
traces. Ideally, each message gets a unique and distinct creation context
assigned. However, as a context must refer to a span this would require the
creation of a distinct span for each message, which is not feasible in all
scenarios. In certain batching scenarios where many messages are created and
published in large batches, creating a span for every single message would
obfuscate traces and is not desirable. Thus having a unique and distinct

> context per message is recommended, but not required.

Author: This was discussed in today's messaging workgroup, the proposed change was applied.

#### Instrumenting pull-based scenarios

In pull-based consumer scenarios, the delivery of messages is requested by the
application code. This usually involves a blocking call, which returns zero or
more messages on completion.

A "Receive" span covers such calls and should link to the "Create" spans of all
messages that are forwarded via the respective call. Depending on the use case,
"Receive" spans can correlate with "Process" spans or other spans modelling
processing operations.

#### General considerations for both push-based and pull-based scenarios

The operations modelled by "Deliver" or "Receive" spans do not strictly refer
to receiving the message from intermediaries, but instead refer to the
application receiving messages for processing. If messages are fetched from the
intermediary and forwarded to the application in one go, the whole operation
might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
might pre-fetch messages from intermediaries and cache those messages, and only
forward messages to the application at a later time. In this case, the
operation of pre-fetching and caching should not be covered by the "Deliver" or

> "Receive" spans.

Reviewer 2: I don't get what it actually means. Is the pre-fetching an implementation detail and an SDK optimization but in the end a message is anyway delivered or received? I mean when the SDK pre-fetches the messages and put them into a cache, no spans are created but when they are provided to the application, the "Deliver" or "Receiver" will be created. Is my assumption correct?

Author: > I mean when the SDK pre-fetches the messages and put them into a cache, no spans are created but when they are provided to the application, the "Deliver" or "Receiver" will be created. Is my assumption correct?

Yes, mostly.

SDKs can surely create spans that cover implementation specific details, but we don't intend to cover this with v1 of these conventions. This might be covered by transport-level instrumentation (e. g. AMQP instrumentation).

Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
to notify the application that no message is available for processing), one
message, or multiple messages (a batch of messages). "Deliver" and "Receive"

> spans should link to the "Create" or "Publish" span of the messages forwarded,

Reviewer 2: If I got right your doubt I think that linking makes more sense when it's not just one consumer getting the same message. The Azure SDK issue you are referring to is related to Service Bus using a queue right? So I can see just one consumer getting the produced message and having everything in one trace.
Different if using Event Hub or Kafka where different consumers (in different consumer groups) could get same produced message, so the need for a different trace linked to the producer one.
How do you get one trace in this case? The producer side should be part of different traces related to different consumers.

> spans should link to the "Create" or "Publish" span of the messages forwarded,

Author: The reasoning to use links exclusively for producer/consumer relationships was the fact that it is modelling an asynchronous relationship (where parent/child relationship are usually synchronous).

You could have a producer trace for publishing a single message, but the message is only consumed hours or days later. Or if there are multiple non-batch consumers, each of those will end up as part of the producer trace.

There definitely are cases where parent/child relationships are the more intuitive solution for end users. However, there are also cases where the results might be confusing.

Anyway, let's discuss it in the next workgroup meeting.

> spans should link to the "Create" or "Publish" span of the messages forwarded,

Reviewer 5: >  Different if using Event Hub or Kafka where different consumers (in different consumer groups) could get same produced message, so the need for a different trace linked to the producer one.
> How do you get one trace in this case? The producer side should be part of different traces related to different consumers.

Each consumer would create a new span that would belong to the same trace - they are be two different branches in trace tree. I don't see any problem with it.

> You could have a producer trace for publishing a single message, but the message is only consumed hours or days later. Or if there are multiple non-batch consumers, each of those will end up as part of the producer trace.

There is no requirement for traces to be short - still, I don't see a problem here. If you'd like to optimize tooling for tail-based sampling or other scenarios like that - long traces don't really impose problems - you'd just buffer parts of these traces at different times. Note that regardless of link/parent choice here, users would create individual processing spans and would prefer using parent-child relationships. I.e. there is never a guarantee of a short trace.

> spans should link to the "Create" or "Publish" span of the messages forwarded,

Author: > There is no requirement for traces to be short - still, I don't see a problem here.

It's not a requirement, it's just that while some customers want both producer and consumer operations in the same trace, that doesn't mean that all (or most) customer will be happy with such a solution.

For the record, I remembered some other reasons why we initially went with links exclusively:
* Consistency over all messaging scenarios. Even if we'd recommend parent/child relationships for single-message scenarios, that might not be possible in pull-based consumer scenarios. If a user calls a `Receive` method to receive message, this method might already have a parent context (in addition to the creation context of the message).
* Keeping possibilities open for intermediary instrumentation. If there's a parent/child relationship between producer and consumer spans, it's not clear how intermediary instrumentation can fit into the picture. It's not impossible, but I think it will be more cumbersome.

However, in my opinion instrumentors should still be allowed to use parent/child relationships for single-message scenarios when it's possible and when it makes sense for them, without breaking compatibility with the semantic conventions (we're not using any `MUST` statements currently). It's not clear to me though whether that's something we should recommend.

I put this up as a discussion point for the next workgroup meeting, I'd love to hear other people's thoughts.

> spans should link to the "Create" or "Publish" span of the messages forwarded,

Author: This was discussed in today's workgroup meeting.

There's consensus that for some single-message scenarios users prefer parent/child relationships between producer and consumer spans instead of links.  However, using parent/child relationships makes it impossible to make a consumer span a child of an ambient context, instead it needs to be a service root span.

We agreed on strongly recommending using links for connecting producer and consumer spans. However, a parent/child relationship may be used in addition to the link in scenarios where this makes sense. With this solution, backends can consistently rely on links, and instrumentations can provide customers with single-trace solutions for scenarios where it makes sense.

> spans should link to the "Create" or "Publish" span of the messages forwarded,

Author: I added this additional text and examples to the OTEP.

> spans should link to the "Create" or "Publish" span of the messages forwarded,

Reviewer 4: > We agreed on strongly recommending using links for connecting producer and consumer spans.

I don't agree with this.  I think the parent/child relationship should be associated with whichever has the strongest affinity. In some cases it might be the producer span in other cases the ambient span. As such, I like the proposal to use parent/child when just a single message is consumed.  I would like to extend that proposal slightly:

Parent/child relationship should be used with the remote span instead of the ambient span if the following are true:
-  the consumed batch contains a single message OR all messages share the same trace id OR there is no ambient span OR a setting is configured for the instrumentation to prefer this behavior.
- (if possible to determine) the message was created/produced within a configurable threshold.

In either case, the span representing the receive/deliver/handle of an individual message should be put in scope for the duration that the message is being processed to signify the handling of that message.

The reason to use a configurable timeout as an additional signal for determining the relationship is that these situations likely represent message reprocessing and much less strongly coupled to the original producer.  In many tracing systems, gaps of long periods of time also are not handled well and visually distort the trace graph. I understand that creation date is not always available to all instrumentation which is why I added "if possible".

> spans can link to zero, one, or multiple "Create" spans.

Reviewer 2: "....hose spans can link to zero...." does it refer to the case when no "Create" span is created but there is just a "Publish" span as mentioned in the "Producer" section? In this case why "Deliver" and "Receive" are not linked to the "Publish" span? Or at least is this mention missing here?
When there is no "Create" span, how we correlate the consumer to the producer?

Author: > "....hose spans can link to zero...." does it refer to the case when no "Create" span is created but there is just a "Publish" span as mentioned in the "Producer" section?

This is to cover the scenario where an application requests messages, but no messages are available for delivery. This results in a "Deliver" or "Receive" operation that refer to zero messages.

Reviewer 2: Could we clarify it just a bit ... even something like " ... can link to zero (no messages received), one, ...." ?

> #### Settlement

Reviewer 2: Shouldn't we clarify that we are talking about consumer settlement here? Saying something like "was manually triggered by the user" sounds like it's just the consumer part settling the received message.
Anyway we should add a producer settlement section or making the following one general enough.

Author: Let's continue this discussion about consumer settlement in another thread: https://github.com/open-telemetry/oteps/pull/220#discussion_r993966940

Messages can be settled in a variety of different ways. In some cases, messages
are not settled at all (fire-and-forget), in other cases settlement operations
are triggered manually by the user, and in callback scenarios settlement can be
automatically triggered by messaging SDKs based on return values of callbacks.

A "Settle" span should be created for every settlement operation, no matter if
the settlement operation was manually triggered by the user or automatically
triggered by SDKs. SDKs will in some cases auto-settle messages in
push-scenarios when messages are delivered via callbacks. In this case, it is

> recommended to create a parent span, so that a "Settle" span will be a sibling

Author: Change the wording: _In cases where it is possible, it is recommended to create the "Settle" span in the scope of the "Deliver" span._

of the related "Deliver" span.

Alternatively, an event can be created instead of a "Settle" span. Events could
be added to "Deliver" spans or to ambient spans.

> "Settle" spans may link to "Create" spans of the messages that are settled,

Author: @lmolkova How about this wording:

> For [individual settlement](https://github.com/open-telemetry/oteps/blob/9a71f85213c325f52bd30b3bef1c5aeafe6dcfbd/text/trace/0173-messaging-semantic-conventions.md#individual-settlement) scenarios, "Settle" spans should link to "Create" spans of the messages that are settled. For [checkpoint based settlement](https://github.com/open-telemetry/oteps/blob/9a71f85213c325f52bd30b3bef1c5aeafe6dcfbd/text/trace/0173-messaging-semantic-conventions.md#checkpoint-based-settlement) this is often not feasible, as it's not known which messages are settled by advancing a checkpoint.

> "Settle" spans may link to "Create" spans of the messages that are settled,

Reviewer 5: even with the individual settlement, message might not be available. For  example,  RabbitMQ settlement in Java looks like [`channel.basicAck(deliveryTag, false);`](https://rabbitmq.github.io/rabbitmq-java-client/api/current/com/rabbitmq/client/Channel.html#basicAck(long,boolean)), so I don't think we need to be specific here and can just say that links should be created when possible.

> "Settle" spans may link to "Create" spans of the messages that are settled,

Author: I'm still wondering whether this should be "SHOULD". I'm thinking e. g. about the case of having a "Settle" span as a child of a "Deliver" span, and both linking to the same "Create" context, and I'm wondering if it's not a bit of an overload to have two links there.

However, I'll make the change as proposed.

however, for some settlement scenarios this is not feasible or possible.

## Proposed changes and additions to the messaging semantic conventions

### Operation name

The following operations related to messages are covered by these semantic
conventions:

| Operation name | Description |
|----------------|-------------|
| `publish`      | One or more messages are provided for publishing to an intermediary. |
| `create`       | A message is created. |
| `receive`      | One or more messages are requested by a consumer. |
| `deliver`      | One or more messages are passed to a consumer. |
| `settle`       | One or more messages are settled. |

For further details about each of those operations refer to the [section about trace structure](#trace-structure).

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|

> | `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |

Reviewer 2: can you elaborate why it's `INTERNAL` if "Create" spans are present please?

Author: If both "publish" and "create" spans are present, and both would be of kind `PRODUCER`, then one would end up with nested producer spans. This is something we want to avoid, as it creates ambiguity and makes it harder to analyze traces.

In addition to avoid nested span, we can also make this statement:

> Setting span kinds according to this table ensures, that span links between consumers and producers always go from a `PRODUCER` span on the producer side to a `CONSUMER` span on the consumer side. This allows analysis tools to interpret linked traces without the need for additional semantic hints.

Having a single producer span for each message, and having a single consumer span for each time a message is received from an intermediary makes it much easier to interpret traces.

> | `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |

Author: In today's workgroup discussion we agreed to keep the "publish" span of kind `PRODUCER` if no "create" spans are present, as in this case it's the context of the "publish" span that will be propagated with the message and linked by consumer spans.

However, if both "create" and "publish" spans are present, there have been different opinions of how to handle the kind of the "publish" span.

```suggestion
| `publish`      | `PRODUCER`, if no `create` spans are present. |
```

One option would be to not specify any span kind for this case and leave it to the instrumentor to decide. Instrumentors should then follow general guidelines in the specification about what span kind to use. Different kinds might be suitable for different scenarios.

```suggestion
| `publish`      | `PRODUCER` |
```

Another option would be to also make the "publish" span of kind `PRODUCER`, which would result in nested `PRODUCER` spans if "create" span are present too.

I tend more towards the first option, as it's in the spirit of just spec'ing out a minimum of requirements. It's also not clear to me what benefits we'd get from requiring "publish" spans to always be of kind `PRODUCER`, but I'd like to hear other opinions.

@joaopgrassi @lmolkova

> | `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |

Reviewer 5: based on [producer kind definition](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind), it indicates start of async request. 

When `Create` spans are recorded, `Publish` span is a simple sync operation and cannot have producer kind.
I suggest we leave kind undefined, we don't have to be that specific in OTEP.

| `create`       | `PRODUCER` |
| `receive`      | `CONSUMER` |
| `deliver`      | `CONSUMER` |

> | `settle`       | `INTERNAL` |

Author: This could also be a `CLIENT` span, as it models a synchronous call to the intermediary.

Remove it from the table and write it as text.

> | `settle`       | `INTERNAL` |

Author: The `settle` operation was removed from the table, a note was added that the kind for `settle` spans should be set according to generic specification about span kinds.

Setting span kinds according to this table ensures, that span links between
consumers and producers always go from a `PRODUCER` span on the producer side to
a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
linked traces without the need for additional semantic hints.

### Span relationships

#### Producer spans

"Publish" spans SHOULD be created for operations of providing messages for
sending or publishing to an intermediary. A single "Publish" span can account
for a single message, or for multiple messages (in the case of providing
messages in batches). "Create" spans MAY be created. A single "Create" span
SHOULD account only for a single message.

If a "Create" span exists for a message, its context SHOULD be attached to
the message. If no "Create" span exists, the context of the related "Publish"

> span SHOULD be attached to the message.

Author: Consensus in the workgroup: if create spans are not children of publish spans, they should be linked.

#### Consumer spans

##### Push-based scenarios

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application
code.

##### Pull-based scenarios

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code.

##### General considerations

"Deliver" or "Receive" spans MUST NOT be created for messages which are not
forwarded to the application, but are pre-fetched or cached by messaging
libraries or SDKs.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received).  For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to

> the "Create" span for the message.

Reviewer 2: or the "Publish" span if the "Create"(s) span(s) don't exist (as we show in the first example below)

Author: Good catch, I added this.

#### Settlement spans

"Settle" spans or events SHOULD be created for every manually or automatically
triggered settlement operation. A single "Settle" span can account for a
single message or for multiple messages (in case messages are passed for
settling as batches). For each message it accounts for, the "Settle" span
MAY link to the "Create" span for the message.

> ## Examples

Reviewer 5: it would be great to somehow identify which spans are expected to be created by auto-instrumentations and which we expect from users. Or maybe even remove process spans from examples since we don't spec them out.

Author: All spans that are not required to satisfy these conventions and are not supposed to created by auto-instrumentation are displayed with increased opacity. All required spans that should be created by auto-instrumentation are created in green.

### Single message producer, single message push-based consumer

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end

>   PM1-. link .->DM1;

Author: I added examples that have both parent/child and link relationships now.

```

### Single message producer, single message push-based consumer with manual settlement

```mermaid
flowchart LR;
  subgraph PRODUCER
    direction TB
    PM1[Publish m1]
  end
  subgraph CONSUMER
    direction TB
    DM1[Deliver m1]-->S1[Settle m1]
  end
  PM1-. link .->DM1;
```

### Single message producer, single message push-based consumer with auto-settlement

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  PM1[Publish m1]
  end
  subgraph CONSUMER
  direction TB

>   A[Ambient]-->DM1[Deliver m1]

Author: The example was adapted, the deliver span is now the parent of the settle span.

  A-->S1[Settle m1]
  end
  PM1-. link .->DM1;
```

### Batch message producer with "Create" spans, single message pull-based consumer

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Publish]-->CM1[Create m1]
  P-->CM2[Create m2]
  end
  subgraph CONSUMER
  direction TB
  RM1[Receive m1]
  RM2[Receive m1]
  end
  CM1-. link .->RM1;
  CM2-. link .->RM2;
```

### Batch message producer, single message push-based consumer

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Publish]
  end
  subgraph CONSUMER
  direction TB
  DM1[Deliver m1]
  DM2[Deliver m1]
  end
  P-. link .->DM1;
  P-. link .->DM2;
```

### Batch message producer with "Create" spans populated before publish, single message pull-based consumer

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  A[Ambient]-->CM1[Create m1]
  A-->CM2[Create m2]
  A-->P[Publish]
  end
  subgraph CONSUMER
  direction TB
  RM1[Receive m1]
  RM2[Receive m1]
  end
  CM1-. link .->RM1;
  CM2-. link .->RM2;

> ```

Reviewer 2: shoud this show that "Create" spans are CHILD_OF the "Publish" one?

Author: No, as the header for this example says: _Batch message producer with "Create" spans populated **before** publish_.

This is to support scenarios where users want full control and provide a creation context on their own, which is then attached to the message.

Reviewer 2: I still don't get it. From the diagram it anyway seems that the consumer has received the messages so there is a link between "Create" and "Receive" spans but "Publish" is not correlated with the "Create" yet which looks really strange. If you are referring to something like " ... spans populated before publish." then I would expect that if not published, we should not have "Receive" spans on the right because not published, means not received yet.

### Single message producers, batch push-based consumer with process spans

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  PM1[Publish m1]
  PM2[Publish m2]
  end
  subgraph CONSUMER
  direction TB
  D[Deliver]-->PRM1[Process m1]
  D-->PRM2[Process m2]
  end
  PM1-. link .->D;
  PM2-. link .->D;
```

### Single message producers, batch pull-based consumer with process spans

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  PM1[Publish m1]
  PM2[Publish m2]
  end
  subgraph CONSUMER
  direction TB

>   A[Ambient]-->R[Receive]

Reviewer 5: I'm not sure what ambient means in this case, but I don't believe it's possible to have ambient context here (created by instrumentation) and in other pull-based examples

dedicated processing operation cannot be identified, or where processing
happens in a different trace. Furthermore, processing operations often are not
covered by messaging libraries and SDKs, but take place in application code.
Consistently creating spans for "Processing" operations would require either
effort from the application owner to correctly instrument those operations, or
additional capabilities of messaging libraries and SDKs (e. g. hooks for
processing callbacks, which can then be instrumented by the libraries or SDKs).

While it is possible to create "Process" spans and correlate those with
consumer traces in certain cases, this is not something that can be generally
required. Therefore, it is more feasible to require the creation of "Deliver"
spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
correlate producer with consumer traces.

#### Instrumenting push-based scenarios

In push-based consumer scenarios, the delivery of messages is not initiated by
the application code. Instead, callbacks or handlers are registered and then
called by messaging SDKs to forward messages to the application.

> A "Deliver" span covers the call of such a callback or handler and should link

Reviewer 5: Does not block otep, but I think we need to get an agreement on this issue (no context in the message)  https://github.com/open-telemetry/opentelemetry-specification/issues/2176

Author: Add this as part of an "Open issues" section to the OTEP.

spans of all messages that are forwarded via the respective call. 

#### General considerations for both push-based and pull-based scenarios

The operations modelled by "Deliver" or "Receive" spans do not strictly refer
to receiving the message from intermediaries, but instead refer to the
application receiving messages for processing. If messages are fetched from the
intermediary and forwarded to the application in one go, the whole operation
might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
might pre-fetch messages from intermediaries and cache those messages, and only
forward messages to the application at a later time. In this case, the
operation of pre-fetching and caching should not be covered by the "Deliver" or
"Receive" spans.

Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
to notify the application that no message is available for processing), one
message, or multiple messages (a batch of messages). "Deliver" and "Receive"
spans should link to the "Create" or "Publish" span of the messages forwarded,
thus those spans can link to zero, one, or multiple "Create" spans.

> If the operation covered by "Deliver" or "Receive" forwards exactly one

Author: Added as suggested: 

> For single-message scenarios, in addition to adding the "Create" span as a link on the "Deliver" or "Receive" span, > the "Create" span may also be used as a parent on those operations. Keeping single-messages operations
in the same trace can greatly improve the user experience.

message, in addition to creating a link to the "Create" span, the "Create" span
may also be used as a parent of the "Deliver" or "Receive" span. While this is
not possible for all scenarios (e. g. when receiving a batch of messages, or
when parenting "Deliver" or "Receive" spans to an ambient context), it can
improve the user experience in some scenarios.

#### Settlement

> Messages can be settled in a variety of different ways. In some cases, messages

Author: Fixed.

are not settled at all (fire-and-forget), in other cases settlement operations
are triggered manually by the user, and in callback scenarios settlement can be
automatically triggered by messaging SDKs based on return values of callbacks.

A "Settle" span should be created for every settlement operation, no matter if
the settlement operation was manually triggered by the user or automatically
triggered by SDKs. SDKs will in some cases auto-settle messages in
push-scenarios when messages are delivered via callbacks. In cases where it is

> possible, it is recommended to create the "Settle" span in the scope of the

Author: Fixed.

| `create`       | A message is created. |
| `receive`      | One or more messages are requested by a consumer. |
| `deliver`      | One or more messages are passed to a consumer. |
| `settle`       | One or more messages are settled. |

For further details about each of those operations refer to the [section about trace structure](#trace-structure).

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|
| `publish`      | `PRODUCER`, if no `create` spans are present. |
| `create`       | `PRODUCER` |
| `receive`      | `CONSUMER` |
| `deliver`      | `CONSUMER` |

The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
e. g. it should be to `CLIENT` if the `settle` spans models a synchronous call
to the intermediary.

> Setting span kinds according to this table ensures, that span links between

Reviewer 6: ```suggestion
Setting span kinds according to this table ensures that span links between
```
I'm not a native English speaker, but the comma here feels strange.

messages in batches). "Create" spans MAY be created. A single "Create" span
SHOULD account only for a single message.

If a "Create" span exists for a message, its context SHOULD be attached to
the message. If no "Create" span exists, the context of the related "Publish"
span SHOULD be attached to the message.

#### Consumer spans

##### Push-based scenarios

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application
code.

##### Pull-based scenarios

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code.

##### General considerations

"Deliver" or "Receive" spans MUST NOT be created for messages which are not
forwarded to the caller, but are pre-fetched or cached by messaging
libraries or SDKs.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received).  For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to

> the "Create" or "Publish" span for the message. In addition, if it is possible

Reviewer 6: If we use the "Publish" span as the "create" context of all messages, then here we will be essentially adding duplicated links, right? Is this something we discussed at some point?

Author: In the context of publishing batches, this will result in having the same context for different messages. I think that's what you mean.

We discussed this, but I can make it more explicit in the text if needed.

Reviewer 6: > In the context of publishing batches, this will result in having the same context for different messages. I think that's what you mean.

Exactly. If you read it in full it becomes clear. My point was more regarding adding links that all point to the same thing. But I guess there's no easy way around that. We can discuss during the meeting, but I fear there's no way around it.

> the "Create" or "Publish" span for the message. In addition, if it is possible

Reviewer 6: Should we phrase this part in a way that it "favors" or give "precedence" for adding it as a parent when possible, and Links being the "fallback" if not? Because (IIRC) that's what Tyler was advocating for and I think it's what a lot of users would prefer. 

In my interpretation the way it is now, it seems the link option takes precedence (or it is more important?) than the second. 

Also: Should we spec something like: If added as parent, then don't add links? Or do we always want the links there for consistency? Thinking as a user consuming a trace, wouldn't it be awkward that it's all in the same trace, and there's a link to "itself"?

Author: > In my interpretation the way it is now, it seems the link option takes precedence (or it is more important?) than the second.

Yes, it's my intention to convey this. In my opinion, the link should always be there. This gives consistency across all scenarios, it's something backends can rely on, and it allows to integrate intermediary instrumentation.

Reviewer 6: I agree it makes things more consistent if we always add links. But I think people consuming these conventions would benefit of having a clear message that parents are still the "desired outcome" _when_ it's possible to do so.

Reviewer 6: I will mark this as resolved, as we already say here:

>In addition, if it is possible the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or "Receive" span.

the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or
"Receive" span.

#### Settlement spans

"Settle" spans or events SHOULD be created for every manually or automatically
triggered settlement operation. A single "Settle" span can account for a
single message or for multiple messages (in case messages are passed for
settling as batches). For each message it accounts for, the "Settle" span

> MAY link to the "Create" or "Publish" span for the message.

Reviewer 6: ```suggestion
MAY link to the "Create" or "Publish" span of the message.
```

I guess?

## Examples

This section contains a list of examples illustrating the use of the
conventions outlined above. Green boxes denote spans that are required to
exist in order to conform to those conventions. Other boxes denote spans that
are not required and covered by the conventions, but are hopefully helpful in

> understanding how messaging spans can be integrated into an overall trace flow.

Reviewer 6: Should we also indicate that doted lines are links and straight lines are parent-child? I fear one looking at this: 
![image](https://user-images.githubusercontent.com/5938087/225966353-de65c898-6a2a-4620-833c-e203c5fd9a0e.png)
Will not understand that the lower one is referring to a parent-child + links scenario.

Author: I added some explanatory text to the introduction section of the examples.

> ### Single message producer, single message push-based consumer

Reviewer 6: Do you think it would make sense in separating the diagrams in "push-based" and "pull-based" sub-sections? Right now 1 they are mixed which makes it hard to find cases and 2 we repeat "push/pull-based" in all diagrams, which makes the diagram title long.

Author: I did so, also replaced headers for individual cases with a more verbose description.

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  PM1-. link .->DM1;

  classDef normal fill:green
  class PM1,DM1 normal
  linkStyle 0 color:green,stroke:green
```

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  PM1-. link .->DM1;
  PM1-->DM1;

  classDef normal fill:green
  class PM1,DM1 normal
  linkStyle 0,1 color:green,stroke:green
```

### Single message producer, single message push-based consumer with manual settlement

```mermaid
flowchart LR;
  subgraph PRODUCER
    direction TB
    PM1[Publish m1]
  end
  subgraph CONSUMER
    direction TB
    DM1[Deliver m1]-->S1[Settle m1]
  end
  PM1-. link .->DM1;

  classDef normal fill:green
  class PM1,DM1,S1 normal
  linkStyle 0,1 color:green,stroke:green
```

> ### Single message producer, single message push-based consumer with auto-settlement

Reviewer 6: Since these are the same, maybe we could combine both `###` sections under the same diagram?

Author: Done.

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Publish]-->CM1[Create m1]
  P-->CM2[Create m2]
  end
  subgraph CONSUMER
  direction TB
  RM1[Receive m1]
  RM2[Receive m2]
  end
  CM1-. link .->RM1;
  CM2-. link .->RM2;

  classDef normal fill:green
  class P,CM1,CM2,RM1,RM2 normal
  linkStyle 0,1,2,3 color:green,stroke:green
```

### Batch message producer, single message push-based consumer

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Publish]
  end
  subgraph CONSUMER
  direction TB
  DM1[Deliver m1]
  DM2[Deliver m2]
  end
  P-. link .->DM1;
  P-. link .->DM2;

  classDef normal fill:green
  class P,DM1,DM2 normal
  linkStyle 0,1 color:green,stroke:green
```

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Publish]
  end
  subgraph CONSUMER
  direction TB
  DM1[Deliver m1]
  DM2[Deliver m2]
  end
  P-. link .->DM1;
  P-. link .->DM2;
  P-->DM1;
  P-->DM2;

  classDef normal fill:green
  class P,DM1,DM2 normal
  linkStyle 0,1,2,3 color:green,stroke:green
```

### Batch message producer with "Create" spans populated before publish, single message pull-based consumer

```mermaid

> flowchart LR;

Reviewer 6: I guess if we go with @lmolkova 's suggestion, we need a link on publish for all the creates, right?

Author: Added.

If a "Create" span exists for a message, its context SHOULD be injected into
the message. If no "Create" span exists, the context of the related "Publish"
span SHOULD be injected into the message.

#### Consumer spans

##### Push-based scenarios

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application
code.

##### Pull-based scenarios

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code.

##### General considerations

"Deliver" or "Receive" spans MUST NOT be created for messages which are not
forwarded to the caller, but are pre-fetched or cached by messaging
libraries or SDKs.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received).  For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
the "Create" or "Publish" span for the message. In addition, if it is possible

> the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or

Reviewer: Can we clarify this for batch-receive, where multiple messages/producer spans "fan in" to a single receive span? There was some dicussion around that for AWS Lambda SQS at https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1051#issuecomment-1492251707

IMHO: For a batch receive, you may still want to set any of the producer spans as parent e.g.:
* One may argue that the last message in the batch is the one that caused(!) the receive operation to finish since it filled up the batch.
* Semantically, links may imply a much weaker causal relationship than parent-child relationships. But for message processing, if not for batches, parent-child would often be quite appropriate.
* For technical reasons, it may be helpful to also have a parent: Links are usually not as convenient to use and process in backends. For example, links may be shown as actual clickable links only, while parent-childs may be shown in a single tree. With parent-child also for batches, you will be able to find an exemplar (parent-child) trace for any particular path through your system given enough samples (i.e. traces).

Author: What you describe might theoretically work for "deliver" cases (push-based scenarios), as in those cases the batch of messages is already ready before the "deliver" span is started. Disregarding internals like message pre-fetching and caching, it's not possible for "receive" cases (pull-based scenarios). In those cases, we must assume that no message is present when the "receive" operation (and span) starts, and messages are fetched during this "receive" operation. Thus, no parent producer context is available and can be used when the "receive" span starts.

> * One may argue that the last message in the batch is the one that caused(!) the receive operation to finish since it filled up the batch.

When talking about causality in the context of parent/child relationships, this usually implies causation for _starting_ an operation, but not for _finishing_ it.

> * Semantically, links may imply a much weaker (non-)causal relationship than parent-child relationships. But for message processing, if not for batches, parent-child would often be quite appropriate.

For single-message cases, this OTEP already recommends creating a parent/child relationship.

> * For technical reasons, it may be helpful to also have a parent: Links are usually not as convenient to use and process in backends. For example, links may be shown as actual clickable links only, while parent-childs may be shown in a single tree. With parent-child also for batches, you will be able to find an exemplar (parent-child) trace for any particular path through your system given enough samples (i.e. traces).

This OTEP doesn't forbid instrumentors to select any message creation context as parent in batch consumer scenarios. If this is considered beneficial for certain scenarios (e. g. SQS), it can be done while still being conforming with this proposal.

However, I think it shouldn't be generally recommended or mandated for all cases. There are many sources of ambiguity and possible confusion that this could cause:
* How would you interpret the trace duration in such a case? You'd end up with a trace that at some points starts with producing message A, but then also includes the time needed for batch-consuming multiple other, possibly unrelated messages. The trace duration doesn't tell you anything anymore about message A: maybe message A is processed very quickly, but another message in the batch takes very long to process?
* Interpreting errors in traces becomes trickier too. You might have a trace started with producing message A, but an error could happen during batch-consuming of another message.
* Also, often consuming message cause others to be produced. You might start out with producing message A, a batch-consuming operation involving message A and message B is parented, during consuming message B, message C is produced, and a batch-consuming operation involving message C is parented to the trace that started with producing message A, although this operation has nothing at all to do with it.

The case of examplar traces is an interesting one, however, when aiming at creating meaningful traces that track a single message, parenting based on a single message in batch-scenarios can be misleading, and even harmful.

Reviewer: You bring up good points. To respond to just one of them: in fact the example I had in mind is probably called a "deliver" span in this spec, despite me using the wrong term "receive".

Reviewer: Re: Trace duration: I wonder if the combination of "span with kind produce" and child has IsRemote (cf. OTEP #182) could/should be used to partition a trace into subtrees for which it is more sensible to calculate duration totals. But I need to read up on how process spans are linked/parented relative to deliver spans...

Author: > "span with kind produce" and child has IsRemote

That could work for some scenarios, but not for batch-consuming ones.

The consistent logic (across all scenarios) with the conventions proposed in this OTEP would be _"span with kind producer" and "linked span(s) of type consumer"_.

Author: @Oberon00 Considering that there is a separate issue that will cover adding semantic conventions for "Process" spans, do you think that this conversation can be closed here and this OTEP unblocked?

from messaging SDKs, without requiring any additional custom instrumentation
from the user.

This OTEP aims at proposing consistent guidelines for creating spans that model
the stages of the messaging flow, and correlating those in a way so that the
requirements sketched above can be met in a consistent way across messaging
scenarios and different messaging systems.

## Terminology

For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).

## Stages of producing and consuming messages

Producing and consuming a message involves five stages:

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  CR[Create] --> PU[Publish]
  end
  subgraph INTERMEDIARY
  direction TB
  end
  subgraph CONSUMER
  direction TB
  RE[Receive] --> PR[Process]
  PR --> SE[Settle]
  end
  PU --> INTERMEDIARY
  INTERMEDIARY --> RE
  SE -..-> INTERMEDIARY
```

1. The producer creates a message.
2. The producer publishes the message to an intermediary.
3. The consumer receives the message from an intermediary.
4. The consumer processes the message.
5. The consumer settles the message by notifying the intermediary that the

>    message was processed. In some cases (fire-and-forget), the settlement stage

Author: Don't mention auto-settlement in this context, as it's ambiugous. But in addition to the fire-and-forget, mention settlement on the broker.

traces. Ideally, each message gets a unique and distinct creation context
assigned. However, as a context must refer to a span this would require the
creation of a distinct span for each message, which is not feasible in all
scenarios. In certain batching scenarios where many messages are created and
published in large batches, creating a span for every single message would
obfuscate traces and is not desirable. Thus instrumentation libraries and
auto-instrumentation should default to creating a unique and distinct context
per message, but may support configuration or other ways to change this default
behavior.

For each producer scenario, a "Publish" span needs to be created. This span
measures the duration of the call or operation that provides messages for
sending or publishing to an intermediary. This call or operation (and the
related "Publish" span) can either refer to a single message or a batch of
multiple messages.

It is recommended to create a "Create" span for every single message. "Create"
spans can be created during the "Publish" operation as children of the
"Publish" span. Alternatively, "Create" spans can be created independently of

> the "Publish" operation. In that case, SDKs may provide mechanisms to allow
> attaching independent contexts with messages.

Author: Correct propagation of context should should not be ensured by messaging SDKs, but by propagators.

```suggestion
the "Publish" operation.
```

If a "Create" span exists for a message, its context must be attached to the
message. If no "Create" span exists for a message, the context of the related

> "Publish" span must be attached to the message.

Author: Replace occurrences of "attach" with "inject".

covered by messaging libraries and SDKs, but take place in application code.
Consistently creating spans for "Processing" operations would require either
effort from the application owner to correctly instrument those operations, or
additional capabilities of messaging libraries and SDKs (e. g. hooks for
processing callbacks, which can then be instrumented by the libraries or SDKs).

While it is possible to create "Process" spans and correlate those with
consumer traces in certain cases, this is not something that can be generally
required. Therefore, it is more feasible to require the creation of "Deliver"
spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
correlate producer with consumer traces.

#### Instrumenting push-based scenarios

In push-based consumer scenarios, the delivery of messages is not initiated by
the application code. Instead, callbacks or handlers are registered and then
called by messaging SDKs to forward messages to the application.

A "Deliver" span covers the call of such a callback or handler and should link
to the "Create" or "Publish" spans of all messages that are forwarded via the

> respective call. Depending on the use case, "Deliver" spans can correlate with
> "Process" spans or other spans modelling processing operations.

Reviewer 5: ```suggestion
respective call. 
```

Motivation:

based on the first sentence ("A "Deliver" span covers the call of such a callback") it sounds like deliver should cover the duration of the callback and likely is a process span. I.e. I can do

`consumer.subscribe(message -> process(message))` and then deliver represents the whole thing.

Or I can do 

`consumer.subscribe(message -> queue.enqueue(message))` and then deliver represents a partial thing with processing happening somewhere else.

Essentially we're not providing any guidance on how to solve it and correlate Process spans, so this sentence that talks about deliver-process correlation seems confusing.

#### Instrumenting pull-based scenarios

In pull-based consumer scenarios, the delivery of messages is requested by the
application code. This usually involves a blocking call, which returns zero or
more messages on completion.

A "Receive" span covers such calls and should link to the "Create" or "Publish"

> spans of all messages that are forwarded via the respective call. Depending on
> the use case, "Receive" spans can correlate with "Process" spans or other spans
> modelling processing operations.

Reviewer 5: ```suggestion
spans of all messages that are forwarded via the respective call. 
```

Same here, when I do this

`messages = consumer.receive(10)`

there is nothing messaging semconv can say and nothing instrumentation libraries can do about how to trace individual messages (or batch after). So the removed sentence does not and cannot provide any guidance anyway

#### General considerations for both push-based and pull-based scenarios

The operations modelled by "Deliver" or "Receive" spans do not strictly refer
to receiving the message from intermediaries, but instead refer to the
application receiving messages for processing. If messages are fetched from the
intermediary and forwarded to the application in one go, the whole operation
might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
might pre-fetch messages from intermediaries and cache those messages, and only
forward messages to the application at a later time. In this case, the
operation of pre-fetching and caching should not be covered by the "Deliver" or
"Receive" spans.

> Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.

Reviewer 5: nit.

maybe

```suggestion
Operations covered by "Receive" can forward zero messages (e. g.
```

I assume deliver callbacks are not called when there is no message?

to notify the application that no message is available for processing), one
message, or multiple messages (a batch of messages). "Deliver" and "Receive"
spans should link to the "Create" or "Publish" span of the messages forwarded,
thus those spans can link to zero, one, or multiple "Create" spans.

If the operation covered by "Deliver" or "Receive" forwards exactly one
message, in addition to creating a link to the "Create" span, the "Create" span
may also be used as a parent of the "Deliver" or "Receive" span. While this is
not possible for all scenarios (e. g. when receiving a batch of messages, or
when parenting "Deliver" or "Receive" spans to an ambient context), it can
improve the user experience in some scenarios.

#### Settlement

Messages can be settled in a variety of different ways. In some cases, messages

> are not settled at all (fire-and-forget), in other cases settlement operations

Reviewer 5: ```suggestion
are not settled at all (fire-and-forget) on happen on the broker, in other cases settlement operations
```

are triggered manually by the user, and in callback scenarios settlement can be
automatically triggered by messaging SDKs based on return values of callbacks.

A "Settle" span should be created for every settlement operation, no matter if
the settlement operation was manually triggered by the user or automatically
triggered by SDKs. SDKs will in some cases auto-settle messages in
push-scenarios when messages are delivered via callbacks. In cases where it is
possible, it is recommended to create the "Settle" span in the scope of the
"Deliver" span.

> Alternatively, an event can be created instead of a "Settle" span. Events could

Reviewer 5: I probably forgot, but why? is there a need to have an event? why not span all the time? how would links work in this case?

Author: If I remember correctly, this was intended for reducing the number of created spans in some scenarios. There can be no links in this case.

Reviewer 5: I'd like to revisit it and would suggest using spans unless there are stronger reasons not to. Rationale:
- deliver spans might be completed by the time settlement happens
- there might be no deliver span (when receive is done)
- there might be no ambient context
- spans don't seem to be much bigger than events in terms of payload, but better supported by tracing backends

Author: Discussed in the workgroup: mentions of events will be removed from the document.

settled, however, for some settlement scenarios this is not feasible or
possible.

## Proposed changes and additions to the messaging semantic conventions

### Operation name

The following operations related to messages are covered by these semantic
conventions:

| Operation name | Description |
|----------------|-------------|
| `publish`      | One or more messages are provided for publishing to an intermediary. |
| `create`       | A message is created. |
| `receive`      | One or more messages are requested by a consumer. |
| `deliver`      | One or more messages are passed to a consumer. |
| `settle`       | One or more messages are settled. |

For further details about each of those operations refer to the [section about trace structure](#trace-structure).

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|
| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
| `create`       | `PRODUCER` |

> | `receive`      | `CONSUMER` |

Reviewer 5: we should adjust span kind definitions on the spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind

currently producer/consumers imply parent-child relationships, which are optional in terms of this semconv

Author: There's an issue submitted for this: https://github.com/open-telemetry/opentelemetry-specification/issues/3172.

Author: Add this as part of an "Open issues" section to the OTEP.

application when those operations are initiated by the application code.

##### General considerations

"Deliver" or "Receive" spans MUST NOT be created for messages which are not
forwarded to the caller, but are pre-fetched or cached by messaging
libraries or SDKs.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received).  For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
the "Create" or "Publish" span for the message. In addition, if it is possible
the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or
"Receive" span.

#### Settlement spans

"Settle" spans or events SHOULD be created for every manually or automatically
triggered settlement operation. A single "Settle" span can account for a
single message or for multiple messages (in case messages are passed for
settling as batches). For each message it accounts for, the "Settle" span
MAY link to the "Create" or "Publish" span for the message.

> 

Reviewer 5: wonder if it would be useful to mention that settle span represents different kinds of settlement: completion (ack), abandoning (nack), dead-lettering, etc

Reviewer 6: I guess that would be interesting to users 🤔 . Would this be then an attribute on the settle span? I'm not that familiar with all but if the majority of "kinds of settlements" are common among messaging systems/sdks then we could add a enum for them. This could be a cool thing for back-ends to use to augment the UI experience, e.g., a nack was received.

Author: I suggest to handle those distinctions via attributes, it should be covered by https://github.com/open-telemetry/opentelemetry-specification/issues/3320.

Author: I linked the above-mentioned issue to this comment. Closing it here.

spans.

## Trace structure

### Producer

Producers are responsible for injecting a creation context into a message.
Subsequent consumers will use this context to link consumer traces to producer
traces. Ideally, each message gets a unique and distinct creation context
assigned. However, as a context must refer to a span this would require the
creation of a distinct span for each message, which is not feasible in all
scenarios. In certain batching scenarios where many messages are created and
published in large batches, creating a span for every single message would
obfuscate traces and is not desirable. Thus instrumentation libraries and
auto-instrumentation should default to creating a unique and distinct context
per message, but may support configuration or other ways to change this default
behavior.

For each producer scenario, a "Publish" span needs to be created. This span
measures the duration of the call or operation that provides messages for
sending or publishing to an intermediary. This call or operation (and the
related "Publish" span) can either refer to a single message or a batch of
multiple messages.

> It is recommended to create a "Create" span for every single message. "Create"

Reviewer 5: I think it's important to outline that if there is existing context on the message (e.g. when it's forwarded), "Create" spans should not need to be created. And "Publish" span should link to the context on the message.

Author: I added this sentence:

> "Create" spans must not be created for forwarded messages into which a context has already been injected.

covered by messaging libraries and SDKs, but take place in application code.
Consistently creating spans for "Processing" operations would require either
effort from the application owner to correctly instrument those operations, or
additional capabilities of messaging libraries and SDKs (e. g. hooks for
processing callbacks, which can then be instrumented by the libraries or SDKs).

While it is possible to create "Process" spans and correlate those with
consumer traces in certain cases, this is not something that can be generally
required. Therefore, it is more feasible to require the creation of "Deliver"
spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
correlate producer with consumer traces.

#### Instrumenting push-based scenarios

In push-based consumer scenarios, the delivery of messages is not initiated by
the application code. Instead, callbacks or handlers are registered and then
called by messaging SDKs to forward messages to the application.

A "Deliver" span covers the call of such a callback or handler and should link
to the "Create" or "Publish" spans of all messages that are forwarded via the

> respective call.

Reviewer 5: Should we express in some way that deliver span covers the status and the duration of the callback and provides ambient parent context for everything inside the callback?

Author: In my understanding, this is implied in "covers the call of such a callback or handler", and I wanted to avoid adding more text and stating the obvious.

However, if you think it adds more clarity I don't mind adding it.

message, or multiple messages (a batch of messages). "Deliver" and "Receive"
spans should link to the "Create" or "Publish" span of the messages forwarded,
thus those spans can link to zero, one, or multiple "Create" spans.

For single-message scenarios, in addition to adding the "Create" span as a link
on the "Deliver" or "Receive" span, the "Create" span may also be used as a
parent on those operations. Keeping single-messages operations in the same
trace can greatly improve the user experience.

#### Settlement

Messages can be settled in a variety of different ways:

* Messages are not settled at all (fire-and-forget).
* Settlement happens on the broker.
* Settlement operations are triggered manually by the user.
* In callback scenarios settlement can be automatically triggered by messaging SDKs
  based on return values of callbacks.

Messages can be settled in a variety of different ways: In some cases, messages
are not settled at all (fire-and-forget), or settlement happens on the broker.
In other cases settlement operations are triggered manually by the user, and in
callback scenarios settlement can be automatically triggered by messaging SDKs

> based on return values of callbacks.

Reviewer 5: nit: this duplicates list above

Author: Fixed.

a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
linked traces without the need for additional semantic hints.

### Span relationships

#### Producer spans

"Publish" spans SHOULD be created for operations of providing messages for
sending or publishing to an intermediary. A single "Publish" span can account
for a single message, or for multiple messages (in the case of providing
messages in batches). "Create" spans MAY be created. A single "Create" span
SHOULD account only for a single message. "Create" spans SHOULD either be
children or links of the related "Publish" span.

If a "Create" span exists for a message, its context SHOULD be injected into
the message. If no "Create" span exists, the context of the related "Publish"
span SHOULD be injected into the message.

#### Consumer spans

##### Push-based scenarios

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application
code.

##### Pull-based scenarios

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code.

> ##### General considerations

Reviewer: It seems the spec skips on "process" spans. I think this is unfortunate, because while it is true that they cannot be "required" and not always produced, in some cases they can, and in probably quite a few cases the application owner might find it worth their while to even add a bit of manual code (maybe just an annotation with appropriate OTel support) as the per-single message processing is vital if you want to know e.g. which messages are taking longer than others to process.

For these cases it would be great if the spec here could also suggest how to set the span relationships.

A process span will often have two potential parents: The "deliver" or "receive" span (or maybe it's parent if such a span is not created) and the (single) Create/Produce span for the message that is processed. So there are three potential ways to capture this (using produce and deliver instead of produce/create and deliver/receive for simplicity):

1. All-links: Link both Produce and Deliver
2. Link Produce and use Deliver as parent (continue trace of Deliver): This emphasizes the technical execution flow of the code, and break the processing of a message (chain) into multiple traces while bundling the processing of the same batch of (potentially unrelated) messages into the same trace.
3. Use Produce as parent and link Deliver (continue trace of Produce): This emphasizes the information flow through the distributed system and will lead to the processing of a (chain of) messages to become a single trace. The technical execution flow will be in linked traces that only go up to the deliver/receive.

I like 3 more from a gut feeling, but there might be cases where 2 or 1 need to be used, e.g. when you have some kind of circular token passing, where a message flow never ends (did include such scenarios in your design considerations?).

Author: > EDIT: It would be OK to create a follow-up issue and mention it in the OTEP text, if you think this process span is particularly tricky.

It's tricky in the sense that we weren't able to come up with conventions that apply to "Process" spans in a general way.

We can generally say that we require a "Receive" spans for all pull-based consumer scenarios, and we can generally say that we require a "Deliver" spans for all push-based consumer scenarios. However, we can't say that we _generally_ require "Process" spans, as those might look different across scenarios. Sometimes they aren't created at all, sometimes there might be one for a batch of messages, sometimes there might be multiple "Process" spans for a single message, sometimes there might be multiple "Process" spans for the same batch ...

The goal of this OTEP is to provide a minimal foundation for instrumenting messaging scenarios, which provides consistency, but gives instrumentors freedom to flexibly add "Process" spans on top of it in ways that makes sense for their scenarios (there's the same aim for transport layer instrumentation).

I think there's limited use in aiming at describing and specifying all the ways in which "Process" spans can be created. Either we artificially limit instrumentor, or we come up with such a long list  that it defies the use of "normative" conventions.

All the three options you describe above are supported with the current draft of conventions in this OTEP (given that "Deliver" links to the "Create" context in all three cases, it can parent in addition). I think that's good and desirable and gives instrumentors flexibility. I think we shouldn't limit them to just one of those options.

Author: Discussed in the workgroup: we will add a follow-up issue and add it as an open issue in the OTEP.

Reviewer: Over-restricting is probably not good here. But for example, the particular question I had in my comment, it might be helpful to add a recommendation at least, i.e. in the form "IF you face the choice between using X and Y as parent, it is conventional but not required to prefer Y.

Author: I submitted a separate issue regarding conventions for "Process" spans: https://github.com/open-telemetry/opentelemetry-specification/issues/3395

I added a reference to it under the "Future possibilities" section of this OTEP, as it's not a blocking issue for the content of this OTEP. However, we intend to start working on it in the very near term.

  to invalid contexts should be allowed and supported.
* [open-telemetry/opentelemetry-specification#3172](https://github.com/open-telemetry/opentelemetry-specification/issues/3172)
  Currently the specification is unclear about whether relationships between
  producer and consumer spans can be modelled via links, the wording suggests
  that it should be a parent/child relationship. The wording in the specification
  needs to make it clear that this can be a link too.

## Examples

This section contains a list of examples illustrating the use of the
conventions outlined above. Green boxes denote spans that are required to
exist in order to conform to those conventions. Other boxes denote spans that
are not required and covered by the conventions, but are hopefully helpful in
understanding how messaging spans can be integrated into an overall trace flow.
Solid arrows denote parent/child relationships, dotted arrows denote link
relationships.

### Push-based scenarios

A producer creates and publishes a single message, the single message is delivered to a consumer:

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end

>   PM1-. link .->DM1;

Reviewer 6: Would you say that the the link arrow should be `Producer < --- Consumer` instead of `Producer -- > Consumer`? I just noticed it now.. I have hacked around mermaid (backwards links [doesn't seem to be supported ](https://github.com/mermaid-js/mermaid/issues/2629#issuecomment-1289228061)) and this code

```
flowchart TB;
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  subgraph PRODUCER
  PM1[Publish m1]
  end
  PM1 --- DM1 .-> |link|PM1
  DM1 --- PM1
  PM1 --- DM1 --> |parent|PM1
  DM1 --- PM1

  classDef normal fill:green
  class PM1,DM1 normal
  linkStyle 0,2,3,5 stroke:none
  linkStyle 1,4 color:green,stroke:green
  ```
  
  Produces this output:
  
  
![image](https://user-images.githubusercontent.com/5938087/230388811-af255084-b66d-422b-b8df-55d5e6a19b0d.png)

I'm also thinking that on the examples where we have "parent", it would be nice to use `top-to-bottom` to emphasize the "parent/child" relationship. Also name the "parent" link.

Reviewer 6: Let me know, I can help out with this as well (perhaps sending a PR to your branch)

Author: I don't have a strong opinion here, if other people too think it's clearer that way, I'd be happy to make the changes.

Personally, I think it's a bit confusing to reverse the direction on the parent and link edges, as edges then would point against the logical flow. For parent/child relationships, a directed edge currently means "is-parent-of", which makes sense in my opinion.

Reviewer 6: With our current design, consumers add link _to_ "producers". Looking at it this way, I just find the current flow in the diagram of `producer --link--> consumer` a bit contradicting. If consumers are the ones adding the link I find it more intuitive that the "arrow" points to the producer instead `producer <--link-- consumer`.

Reviewer 6: I opened a PR with the changes for this. https://github.com/pyohannes/oteps/pull/3. Some diagrams look more "messy" than they looked before. No way to improve the arranging of things in it so I'm not sure. Maybe leaving as-is is the best.

Reviewer 6: I decided to not go ahead with this, as the diagrams look way worse. I did extract the `parent` label on the relationships and now is part of this PR.

batch of message to a consumer:

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  PM1[Publish m1]
  PM2[Publish m2]
  end
  subgraph CONSUMER
  direction TB
  D[Deliver]-.-PRM1[Process m1]
  D-.-PRM2[Process m2]
  end
  PM1-. link .->D;
  PM2-. link .->D;

  classDef normal fill:green
  class PM1,PM2,D normal
  classDef additional opacity:0.4
  class PRM1,PRM2 additional
  linkStyle 0,1 opacity:0.4
  linkStyle 2,3 color:green,stroke:green
```

### Pull-based scenarios

A producer creates and publishes a single message, the single message is
delivered to a consumer. "Create" spans are created independently of the
"Publish" operation:

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  A[Ambient]-->CM1[Create m1]
  A-->CM2[Create m2]
  A-->P[Publish]
  end
  subgraph CONSUMER
  direction TB
  RM1[Receive m1]
  RM2[Receive m2]
  end
  CM1-. link .->RM1;
  CM2-. link .->RM2;
  CM1-. link .->P;
  CM2-. link .->P;

  classDef normal fill:green
  class CM1,CM2,P,RM1,RM2 normal
  classDef additional opacity:0.4
  class A additional
  linkStyle 0,1,2 opacity:0.4
  linkStyle 3,4,5,6 color:green,stroke:green
```

> "Create" spans are created as part of the "Publish" operation:

Reviewer 5: in the description above, publish span above needs to link to create spans, so it seems those should be created first and then publish spans has to link to them. Then they should be siblings (if all happen in the same ambient context)

Author: Isn't that exactly what's depicted in the diagram above? That's the header:

> A producer creates and publishes a single message, the single message is delivered to a consumer. "Create" spans are created independently of the "Publish" operation:

And that's the diagram:
![image](https://user-images.githubusercontent.com/16337442/231582674-403935d9-41f9-4eb2-808e-102479eebc68.png)

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Publish]-->CM1[Create m1]
  P-->CM2[Create m2]
  end
  subgraph CONSUMER
  direction TB
  RM1[Receive m1]
  RM2[Receive m2]
  end
  CM1-. link .->RM1;
  CM2-. link .->RM2;

  classDef normal fill:green
  class P,CM1,CM2,RM1,RM2 normal
  linkStyle 0,1,2,3 color:green,stroke:green
```

A producer creates and publishes a single message, it is delivered as part of a

> batch of message to a consumer. "Process" spans for single messages can be

Reviewer 6: ```suggestion
batch of messages to a consumer. "Process" spans for single messages can be
```

> created, but are not covered by these conventions:

Reviewer 6: Should we also offer a diagram saying that process spans can also be linked to create ones? Also, the "Push" based doesn't have such diagrams, I wonder if should have the same diagrams/cases in both to be consistent? I say because one might think that "process" can only be instrumented in pull based, since there's no example of it in the push-based section.

Author: > Should we also offer a diagram saying that process spans can also be linked to create ones?

No, I'd like to avoid showing those links when they are not part of conventions. I'd like to avoid the situation we have with the current semantic conventions, where people are just inferring things from examples that are otherwise unspecified.

> Also, the "Push" based doesn't have such diagrams, I wonder if should have the same diagrams/cases in both to be consistent? 

I think this diagram covers a similar case for the "Push" based scenarios:
![image](https://user-images.githubusercontent.com/16337442/231586574-de82ecde-f0ad-42b5-8c98-90524f581fbf.png)

Author: @joaopgrassi  Please resolve this conversation if the diagram shown above covers the scenario you had in mind.

Reviewer 6: All good!

4. The consumer processes the message.
5. The consumer settles the message by notifying the intermediary that the
   message was processed. In some cases (fire-and-forget scenarios, or when
   settlement happens on the broker), the settlement stage does not exist.

The semantic conventions described below define how to model those stages with
spans.

## Trace structure

### Producer

Producers are responsible for injecting a creation context into a message.
Subsequent consumers will use this context to link consumer traces to producer
traces. Ideally, each message gets a unique and distinct creation context
assigned. However, as a context must refer to a span this would require the
creation of a distinct span for each message, which is not feasible in all
scenarios. In certain batching scenarios where many messages are created and
published in large batches, creating a span for every single message would
obfuscate traces and is not desirable. Thus instrumentation libraries and
auto-instrumentation should default to creating a unique and distinct context
per message, but may support configuration or other ways to change this default

> behavior.

Author: I added an additional sentence:

> The latter can help to reduce the number of spans and to avoid overly verbose traces.

> For each producer scenario, a "Publish" span needs to be created. This span

Reviewer 7: under which trace id?

Author: The assumption here is, that spans are created as [defined by the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#determining-the-parent-span-from-a-context): use the currently active span as parent (using its trace id), or otherwise start a new trace (with a new trace id).

Author: I guess this refers to the relationship between "Publish" and "Create" spans specifically?

If so, I tried to clarify [here](https://github.com/open-telemetry/oteps/pull/220/files#r1172940271):
* "Create" spans can be children of "Publish" spans. In this they are part of the same trace.
* "Create" spans can be part of the same trace as "Publish" spans without being children of the "Publish" operation. In this case "Publish" spans should link to "Create" spans.
* "Create" and "Publish" spans belong to different traces. In this case "Publish" spans should link to "Create" spans.

measures the duration of the call or operation that provides messages for
sending or publishing to an intermediary. This call or operation (and the
related "Publish" span) can either refer to a single message or a batch of
multiple messages.

It is recommended to create a "Create" span for every single message. "Create"

> spans can be created during the "Publish" operation as children of the

Reviewer 7: this is confusing. If Publish is a reflection of the batching (under which trace id?) and Create spans are children then they will be all under the same trace id.

Author: > "Create" spans can be created during the "Publish" operation as children of the "Publish" span.

In this case, "Create" and "Publish" spans will have the same trace id (that's a requirement for a parent/child relationship).

> Alternatively, "Create" spans can be created independently of the "Publish" operation. In this case, the "Publish" span should link to the "Create" spans.

In this case, "Create" and "Publish" spans can either have the same or different trace ids. In practice, this will likely mean that a user will manually attach a create context to a message _before_ it is passed to the "Publish" operation.

Both scenarios are valid, and we shouldn't prevent or discourage any of them.

"Publish" span. Alternatively, "Create" spans can be created independently of

> the "Publish" operation. In this case, the "Publish" span should link to the

Reviewer 6: ```suggestion
the "Publish" operation (e.g., in some parts of the application, before calling "publish"). In this case, the "Publish" span should link to the
```

Should we be more specific in here, that the spans can be created "before" calling Publish on a messaging library?

Author: I added some additional wording:

> Alternatively, "Create" spans can be created independently of the "Publish" operation, e. g. in cases where messages are created before they are passed to a "Publish" operation.

"Create" spans.

> If a "Create" span exists for a message, its context must be injected into the

Reviewer 6: I'm reading this statement again and I have a few concerns..

> If a "Create" span exists for a message, its context must be injected into the message

If there's a create span, it should already be inside the message, no? 

> if no "Create" span exists for a message, the context of the related "Publish" span must be injected into the message.

This is conflicting with the statement above:

> "Create" spans can be created during the "Publish" operation as children of the "Publish" span

I think then it's unclear what should I do if messages arrive at "publish" without a context

Should instrumentation create a "create" span per message and inject that? OR should instrumentation attach the "publish" span to all messages?

Reviewer 6: Here's one suggestion I thought of:

```
It is recommended to create a "Create" span for every single message. 
"Create" spans can be created during the "Publish" operation as children of the "Publish" span. 
Alternatively, "Create" spans can be created independently of the "Publish" operation 
(e.g., in some other parts of the application, before calling "publish"). 
In this case, the "Publish" span should link to the "Create" spans already present in the message.

Alternatively, if no "Create" span exists for a message, and during publish it is not desired
to create individual spans for each message, 
the context of the related "Publish" span must be injected into the message.
```

Author: I will think a bit more about how to polish this section.

My intention was to have separate paragraphs for (1) defining which spans to create, (2) defining which context to attach. I didn't want to mix up those two to avoid confusion.

Regarding context propagation, the two possibilities are:
1. A "Create" span is created for a message, either as child of "Publish" or independently of it. In any case, its context should be associated with the message.
2. No "Create" span is created. In that case, the context of the "Publish" span should be associated with the message, as this span is required to exist.

> It is recommended to create a "Create" span for every single message. 
"Create" spans can be created during the "Publish" operation as children of the "Publish" span. 
Alternatively, "Create" spans can be created independently of the "Publish" operation 
(e.g., in some other parts of the application, before calling "publish"). 
In this case, the "Publish" span should link to the "Create" spans already present in the message.
>
> Alternatively, if no "Create" span exists for a message, and during publish it is not desired
to create individual spans for each message, 
the context of the related "Publish" span must be injected into the message.

This is a bit confusing to me, as it doesn't explicitly specify what context to attach if a "Create" span is present.

message. If no "Create" span exists for a message, the context of the related
"Publish" span must be injected into the message.

> "Create" spans must not be created for forwarded messages into which a context

Reviewer 6: Can't we simply say here that Create spans must not be created/changed if the message already has one attached? Not sure "forwarded" is a concept that can be know during runtime? We simply want to enforce that it's immutable and once set, it can only be read, but not modified.

Reviewer 6: Suggestion:

```
A "Create" context in a message is immutable. 
Once set, it should not change, thus a "Create" context should 
not be created for messages that already have one set.
```

Author: I'll remove the word "forwarded", I think you're right that it's out of place.

What I'd like to convey in this sentence, is that there can only be one "Create" span for each message. I'd like to avoid talking about replacing context here, I think we made that clear in the context propagation part that's already merged into the spec. I'd like to make clear that for one message, there should only be one "Create" _span_.

With your proposed wording, that wouldn't be clear: it would allow to have additional "Create" spans, as long as the context in the message is not touched.

has already been injected.

### Consumer

For many use cases, it is not possible to rely on the presence of "Process"
spans for correlating producer with consumer traces: there are cases where a
dedicated processing operation cannot be identified, or where processing

> happens in a different trace. Furthermore, processing operations often are not

Reviewer 6: ```suggestion
happens in a different trace. Furthermore, processing operations are not often
```
I'm not a native speaker, but to me this reads better?

| Operation name | Span kind|
|----------------|-------------|
| `publish`      | `PRODUCER`, if no `create` spans are present. |
| `create`       | `PRODUCER` |
| `receive`      | `CONSUMER` |
| `deliver`      | `CONSUMER` |

The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
e. g. it should be set to `CLIENT` if the `settle` spans models a synchronous call
to the intermediary.

Setting span kinds according to this table ensures that span links between
consumers and producers always go from a `PRODUCER` span on the producer side to
a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
linked traces without the need for additional semantic hints.

### Span relationships

#### Producer spans

"Publish" spans SHOULD be created for operations of providing messages for
sending or publishing to an intermediary. A single "Publish" span can account
for a single message, or for multiple messages (in the case of providing

> messages in batches). "Create" spans MAY be created. A single "Create" span

Author: We always expect a "Publish" span to be present, but we allow for scenarios without "Create" spans. So I their requirement levels should be different (`SHOULD` versus `MAY`). I'd like to avoid going to `MUST` and `SHOULD`, for reasons outlined in [another comment](https://github.com/open-telemetry/oteps/pull/220/files#r1184328916).

SHOULD account only for a single message. "Create" spans SHOULD either be
children or links of the related "Publish" span.

> If a "Create" span exists for a message, its context SHOULD be injected into

Author: I'd like to avoid any normative `MUST` wording regarding span structure and context injection. I can think of cases where  attaching certain contexts is not technically possible or feasible, and I think it makes sense in those cases to allow best-effort instrumentations to be compliant.

the message. If no "Create" span exists, the context of the related "Publish"
span SHOULD be injected into the message.

#### Consumer spans

##### Push-based scenarios

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application
code.

##### Pull-based scenarios

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code.

##### General considerations

"Deliver" or "Receive" spans MUST NOT be created for messages which are not
forwarded to the caller, but are pre-fetched or cached by messaging
libraries or SDKs.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received).  For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to

> the "Create" or "Publish" span for the message. In addition, if it is possible

Author: That's the whole sentence:

> For _each_ message it accounts for, the "Deliver" or "Receive" span SHOULD link to the "Create" or "Publish" span for the message.

I think with your proposed change, it would be this:

> For message(s) it accounts for, the "Deliver" or "Receive" span SHOULD link to the "Create" or "Publish" span for the message(s).

I find the last one a bit cumbersome to read, but I'd be happy to make any changes that make things clearer.

> the "Create" or "Publish" span for the message. In addition, if it is possible

Reviewer 6: I was only suggesting for the least portion, but yeah I can see it would make sense to add in all. Let's keep as is, it was just a Nit anyway :)

  needs to make it clear that this can be a link too.

## Examples

This section contains a list of examples illustrating the use of the
conventions outlined above. Green boxes denote spans that are required to
exist in order to conform to those conventions. Other boxes denote spans that
are not required and covered by the conventions, but are hopefully helpful in
understanding how messaging spans can be integrated into an overall trace flow.
Solid arrows denote parent/child relationships, dotted arrows denote link
relationships.

### Push-based scenarios

A producer creates and publishes a single message, the single message is delivered to a consumer:

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  PM1-. link .->DM1;

  classDef normal fill:green
  class PM1,DM1 normal
  linkStyle 0 color:green,stroke:green
```

When consuming a single message, the "Deliver" spans can be parented to the creation context:

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  PM1-. link .->DM1;

>   PM1-->|parent|DM1;

Reviewer 6: Nope, no reason! I will change it to be consistent.

scenarios. In certain batching scenarios where many messages are created and
published in large batches, creating a span for every single message would
obfuscate traces and is not desirable. Thus instrumentation libraries and
auto-instrumentation should default to creating a unique and distinct context
per message, but may support configuration or other ways to change this default
behavior. The latter can help to reduce the number of spans and to avoid overly
verbose traces.

For each producer scenario, a "Publish" span needs to be created. This span
measures the duration of the call or operation that provides messages for
sending or publishing to an intermediary. This call or operation (and the
related "Publish" span) can either refer to a single message or to a batch of
multiple messages.

It is recommended to create a "Create" span for every single message. "Create"
spans can be created during the "Publish" operation as children of the
"Publish" span. Alternatively, "Create" spans can be created independently of
the "Publish" operation, e. g. in cases where messages are created before they
are passed to a "Publish" operation. In this case, the "Publish" span should

> link to the "Create" spans.

Author: I changed the wording, highlighting three possible options:
1. Using a user-specified creation context.
2. Using the "Create" span as creation context.
3. Using the "Publish" span as creation context.

I hope this makes things clearer ...

to receiving the message from intermediaries, but instead refer to the
application receiving messages for processing. If messages are fetched from the
intermediary and forwarded to the application in one go, the whole operation
might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
might pre-fetch messages from intermediaries and cache those messages, and only
forward messages to the application at a later time. In this case, the
operation of pre-fetching and caching should not be covered by the "Deliver" or
"Receive" spans.

Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
to notify the application that no message is available for processing), one
message, or multiple messages (a batch of messages). "Deliver" and "Receive"
spans should link to the "Create" or "Publish" span of the messages forwarded,
thus those spans can link to zero, one, or multiple "Create" spans.

For single-message scenarios, in addition to adding the "Create" span as a link
on the "Deliver" or "Receive" span, the "Create" span may also be used as a
parent on those operations. Keeping single-messages operations in the same

> trace can greatly improve the user experience.

Author: What do you think about the following wording:

> Operations covered by "Deliver" or "Receive" can forward zero messages (e. g. to notify the application that no message is available for processing), one message, or multiple messages (a batch of messages). "Deliver" and "Receive" spans should link to the creation context of the messages forwarded, thus those spans can link to zero, one, or multiple producer spans.
>
> For single-message scenarios, and if the "Deliver" or "Receive" spans would be root spans of a new trace, the creation context may also be used as a parent on those operations in addition to being added as a link. Keeping single-messages operations in the same trace can greatly improve the user experience.

@dpauls

> trace can greatly improve the user experience.

Reviewer 8: This sounds good to me.

#### Settlement

Messages can be settled in a variety of different ways:

* Messages are not settled at all (fire-and-forget).
* Settlement happens on the broker.
* Settlement operations are triggered manually by the user.
* In callback scenarios settlement can be automatically triggered by messaging SDKs
  based on return values of callbacks.

A "Settle" span should be created for every settlement operation, no matter
which party triggered it.  SDKs will, in some cases, auto-settle messages in
push-scenarios when messages are delivered via callbacks. In cases where it is
possible, it is recommended to create the "Settle" span as a child of the

> "Deliver" span.

Author: Proposal after discussions in the WG: don't require a settlement span for fire-and-forget settlement operations.

> "Deliver" span.

Author: Removed the "fire-and-forget" scenario from the list of settlement scenarios, and added a note about settlement spans not being required for this scenario.

| `publish`      | One or more messages are provided for publishing to an intermediary. |
| `create`       | A message is created. |
| `receive`      | One or more messages are requested by a consumer. |
| `deliver`      | One or more messages are passed to a consumer. |
| `settle`       | One or more messages are settled. |

For further details about each of those operations refer to the [section about trace structure](#trace-structure).

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|
| `publish`      | `PRODUCER`, if no `create` spans are present. |
| `create`       | `PRODUCER` |
| `receive`      | `CONSUMER` |
| `deliver`      | `CONSUMER` |

The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
e. g. it should be set to `CLIENT` if the `settle` spans models a synchronous call

> to the intermediary.

Author: Good question, I know initial thoughts were that it can either be a `CLIENT` or an `INTERNAL` span.

You are right, if settlement is done as an asynchronous operations, the wording here would recommend to set use `PRODUCER ` as span kind for the settlement span.

@lmolkova @joaopgrassi, what are your thoughts about that?

> to the intermediary.

Reviewer 6: Yeah `PRODUCER` would be somewhat strange to me as well. Maybe it has to always be `CLIENT`?

Setting span kinds according to this table ensures that span links between
consumers and producers always go from a `PRODUCER` span on the producer side to
a `CONSUMER` span on the consumer side. This allows analysis tools to interpret

> linked traces without the need for additional semantic hints.

Author: The invariant here is, that there always is a `PRODUCER` span on the producer side, which links to a `CONSUMER` span on the consumer side.

Depending on directions we come up with for settlement spans, there might be an additional `PRODUCER` span on the consumer side.

### Span relationships

#### Producer spans

"Publish" spans SHOULD be created for operations of providing messages for
sending or publishing to an intermediary. A single "Publish" span can account
for a single message, or for multiple messages (in the case of providing
messages in batches). "Create" spans MAY be created. A single "Create" span
SHOULD account only for a single message. "Create" spans SHOULD either be
children or links of the related "Publish" span.

If a "Create" span exists for a message, its context SHOULD be injected into
the message. If no "Create" span exists, the context of the related "Publish"

> span SHOULD be injected into the message.

Author: Fixed.

#### Consumer spans

##### Push-based scenarios

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application
code.

##### Pull-based scenarios

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code.

##### General considerations

"Deliver" or "Receive" spans MUST NOT be created for messages which are not
forwarded to the caller, but are pre-fetched or cached by messaging
libraries or SDKs.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received).  For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to

> the "Create" or "Publish" span for the message. In addition, if it is possible

Author: Done.

the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or

> "Receive" span.

Reviewer 8: I don't like leaving the door open to short circuiting the transport trace. The purpose of the link in the first place was to allow for an easy bypass of the transport-level details. This would create an unnatural span structure if transport layer tracing had occurred.

Without getting into details of how transport layer tracing occurs (out of scope for now), my suggestion would be to omit this statement.

Author: Personally I'd agree, however, many people have expressed a strong desire for the possibility to capture simple messaging workflows in a single trace.

> This would create an unnatural span structure if transport layer tracing had occurred.

The worst case I can think of is, when transport layer tracing is enabled on producer-side, but not on consumer-side, and a consumer span is parented to the creation context. Then, both the top-level transport span and the consumer-side "Deliver" or "Receive" span would be set as a parent of the creation context. 

While I don't consider this ideal, I think it's an acceptable trade-off.

#### Settlement spans

"Settle" spans SHOULD be created for every manually or automatically triggered
settlement operation. A single "Settle" span can account for a single message
or for multiple messages (in case messages are passed for settling as batches).
For each message it accounts for, the "Settle" span MAY link to the "Create" or
"Publish" span of the message.

## Open issues

Fully integrating the changes proposed in this document into messaging semantic
conventions requires some additions and clarifications in the specification,
which are listed in this section.

* [open-telemetry/opentelemetry-specification#454](https://github.com/open-telemetry/opentelemetry-specification/issues/454)

>   To instrument push-based "Receive" operations as described in this document,

Reviewer 8: push-based "Deliver" operations?

Author: It should rather read _pull-based "Receive" operations_. Fixed.

  attributes as part of a link which points to an invalid context. However,
  according to the specification links pointing to an invalid context may be
  ignored. For consistently reporting message-specific attributes on links, links
  to invalid contexts should be allowed and supported.
* [open-telemetry/opentelemetry-specification#3172](https://github.com/open-telemetry/opentelemetry-specification/issues/3172)
  Currently the specification is unclear about whether relationships between
  producer and consumer spans can be modelled via links, the wording suggests
  that it should be a parent/child relationship. The wording in the specification
  needs to make it clear that this can be a link too.

## Examples

This section contains a list of examples illustrating the use of the
conventions outlined above. Green boxes denote spans that are required to
exist in order to conform to those conventions. Other boxes denote spans that
are not required and covered by the conventions, but are hopefully helpful in
understanding how messaging spans can be integrated into an overall trace flow.
Solid arrows denote parent/child relationships, dotted arrows denote link

> relationships.

Reviewer 8: I find the direction of the link arrows confusing. While this text doesn't specifically say which span has the link, I would think it is clearest for the base of the arrow to be connected to the span that has the link and the head of the arrow to point at the span the link is referencing. If other find the reverse to be more natural, might it help to define which end of the arrow has the link and which end the link references?

Author: I updated diagrams now to just use dotted lines without arrows for links.

### Push-based scenarios

A producer creates and publishes a single message, the single message is delivered to a consumer:

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  PM1-. link .->DM1;

  classDef normal fill:green
  class PM1,DM1 normal
  linkStyle 0 color:green,stroke:green
```

When consuming a single message, the "Deliver" spans can be parented to the creation context:

```mermaid
flowchart LR;
  subgraph PRODUCER
  PM1[Publish m1]
  end
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  PM1-. link .->DM1;
  PM1-- parent -->DM1;

  classDef normal fill:green
  class PM1,DM1 normal
  linkStyle 0,1 color:green,stroke:green
```

It is recommended to add spans for settlement operations. Those spans can

> either be created manually or via auto-instrumentation:

Reviewer 6: @carlosalberto about your points above:

> Example of auto Settlement span?

This diagram exemplifies settlement spans already, but it's not clear if it's "manual" or "auto" settlement. We agreed to modify the text above the diagram to make it more clear. Do you think that addresses it? Here's a preview of how it would look:

![image](https://github.com/open-telemetry/oteps/assets/5938087/03217de2-3b70-45d7-80e5-e577daeeb257)

```mermaid
flowchart LR;
  subgraph PRODUCER
    direction TB
    PM1[Publish m1]
  end
  subgraph CONSUMER
    direction TB
    DM1[Deliver m1]-->S1[Settle m1]
  end
  PM1-. link .->DM1;
  PM1-. link .->S1;

  classDef normal fill:green
  class PM1,DM1,S1 normal
  linkStyle 0,1,2 color:green,stroke:green
```

> A producer publishes a batch of messages, single messages are delivered to consumers:

Reviewer 8: When publishing a batch, we are saying each message should get a unique creation context, and the deliver/receive spans should link to the creation context. This diagram doesn't show that.

Author: Good point, we probably should endorse recommended practices in the examples we provide. I added "Create" spans in the diagram.

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Publish]
  end
  subgraph CONSUMER
  direction TB
  DM1[Deliver m1]
  DM2[Deliver m2]
  end
  P-. link .->DM1;
  P-. link .->DM2;

  classDef normal fill:green
  class P,DM1,DM2 normal
  linkStyle 0,1 color:green,stroke:green
```

> When consuming a single message, the "Deliver" spans can be parented to the creation context:

Reviewer 8: Again, here we aren't showing each message having a unique creation context.

Author: I added "Create" spans in the example.

spans can link to zero, one, or multiple producer spans.

For single-message scenarios, and if the "Deliver" or "Receive" spans would be
root spans of a new trace, the creation context may also be used as a parent on
those operations in addition to being added as a link. Keeping single-messages
operations in the same trace can greatly improve the user experience.

#### Settlement

Messages can be settled in a variety of different ways:

* Settlement happens on the broker.
* Settlement operations are triggered manually by the user.
* In callback scenarios settlement can be automatically triggered by messaging SDKs
  based on return values of callbacks.

A "Settle" span should be created for every settlement operation, no matter
which party triggered it.  SDKs will, in some cases, auto-settle messages in
push-scenarios when messages are delivered via callbacks. In cases where it is
possible, it is recommended to create the "Settle" span as a child of the
"Deliver" span.

"Settle" spans should link to creation context of the messages that are
settled, when possible.

No settlement span should be created for "fire-and-forget" scenarios, where

> messages aren't settled at all.

Reviewer 8: I've had a number of comments and concerns regarding settlement within the working group meetings. I think it comes down to not clearly defining what we mean by settlement, and properly considering all types of settlement. On the topic of message settlement, I think AMQP does an excellent job of describing the 3 settlement modes that I've seen used by various messaging systems.

There's one small part of one section of the AMQP spec that I think does a good job of summing them up. It is the last part of section 2.6.12, starting with the sentence beginning with "This basic pattern...":
http://docs.oasis-open.org/amqp/core/v1.0/csprd01/amqp-core-transport-v1.0-csprd01.html#:~:text=endpoint%20to%20settle.-,This%20basic%20pattern,-can%20be%20modified

Summarizing even more quickly:
* At-most-once: Sender settles before send, no settlement on receiver.
* At-least-once: Receiver settles after process, settlement is zero duration.
* Exactly-once: Sender settles after receiver transitions a message to terminal state, receiver settles after sender settles. Receiver settlement involves a round trip exchange with the sender.

I think it important for this OTEP to provide recommendations on how to trace settlement for all of these delivery modes. Furthermore, I think it should be possible to discern from the tracing output which delivery mode is being used. Some of this may come from span structure (this OTEP) and some may come from attributes (out of scope for this OTEP), but I think we should understand what our intentions are to ensure we are making the correct recommendations here.

The two easiest are delivery modes that are fairly well described here already, at-most-once and exactly-once. I'll discuss at-least-once last.

### At-most-once Delivery
With at-most-once delivery, the message itself is fire and forget, and there's nothing to record with regards to settlement in the receiver. Since the messaging model presented in the "Stages of Producing and Consuming Messages" shows consumers receive messages from an intermediary, and intermediary instrumentation is out of scope, there is nothing to trace with regards to at-most-once delivery to consumers. I believe this is what this sentence is trying to convey:
> No settlement span should be created for "fire-and-forget" scenarios, where messages aren't settled at all.

My only comment with regards to this sentence is that "messages aren't settled at all" should perhaps say "the receiver" is not involved in settling the message delivery. Ideally "fire-and-forget" would be replaced with "at-most-once delivery":
> No settlement span should be created for at-most-once delivery scenarios, where the receiver is not involved in settling the message.

### Exactly-once Delivery
With exactly once delivery, settlement on the consumer begins when the consumer acknowledges (i.e. transitions message to a terminal state), and ends when the consumer receives confirmation of that acknowledgment from the intermediary. I think this is the delivery mode that is primarily in mind with most discussion of settlement so far in this OTEP.

### At-least-once Delivery
With at-least-once delivery, the receiver settles, and although the settlement itself is a fire-and-forget operation, recording the settlement is important, as it indicates the outcome of processing the message.

There are two scenarios we might consider with at least once delivery settlement:
1. Settlement occurs within the context of an active "Deliver" span (either due to a client ack within the context of the deliver callback, or an "auto ack" based on the callback result). In this case, how should we record settlement? This is a very common scenario, so I think it might be worth considering a special case to optimize. Requiring a zero duration "Settle" span would be unfortunate.
2. In "pull" scenarios, or in "push" scenarios where the message is not processed before the delivery callback returns, I don't see an option other than to create a zero duration "Settle" span to record the result.

Author: > * At-most-once: Sender settles before send, no settlement on receiver.
> * At-least-once: Receiver settles after process, settlement is zero duration.
> * Exactly-once: Sender settles after receiver transitions a message to terminal state, receiver settles after sender settles. Receiver settlement involves a round trip exchange with the sender.

The OTEP proposes to create a "Settle" span for at-least-once and exactly-once scenarios, as those are the scenarios for which any settlement operation happens on the consumer side. I think this conceptually makes sense.

I'm not sure if it is correct to say that "Settle" spans in the at-least-once scenario are zero-duration. It doesn't involve a round-trip exchange, but it involves the consumer sending something over the wire. As with "Create" spans, the duration might in most cases not be very interesting, however, I think it still represents an operation that can be modelled by a span.

That said, I like the terminology that you propose, and I think we should adopt it when talking and writing about settlement. I will rewrite the section accordingly.

However, I don't see substantial changes to the actual span structure we propose, which is summarized in the "Proposed changes and additions to the messaging semantic conventions" section:

> "Settle" spans SHOULD be created for every manually or automatically triggered settlement operation. A single "Settle" span can account for a single message or for multiple messages (in case messages are passed for settling as batches). For each message it accounts for, the "Settle" span MAY link to the creation context of the message.

Reviewer 5: I still don't understand why we'd need to trace settlement (when it happens on broker) on the client.
Perhaps what we want is to record that message was delivered to a client (that we should cover with transport-level instrumentation in the future)? 

When auto-settlement on the broker is configured, this transport-level client-got-a-message event/span would be useful without any settlement semantics and it'd be confusing to have a settlement-specific one  there.

Author: > I still don't understand why we'd need to trace settlement, when it happens on the broker, on the client side.

We don't need to do that. As @dpauls said, "there is nothing to trace with regards to at-most-once delivery to consumers".

I tried to change the text in the settlement section to make this clearer.

Author: Based on discussions in the workgroup meeting: references to the AMQP terminology (at-most-once, at-least-once, exactly-once) were removed to avoid confusion.

processing stages of a message, it should also be possible to determine the
end-to-end latency of processing a message. If intermediaries are not
instrumented, this shouldn't impact the correlation of producer and consumer
stages. If, on the other hand, intermediaries are instrumented, spans from
intermediary instrumentation should seamlessly integrate with producer and
consumer instrumentation. This integration should not require any changes in
producer or consumer instrumentation, and it should not cause any changes to
the relationships of producer and consumer spans. Furthermore, it should be
possible to provide tracing instrumentation as an out-of-the-box experience
from messaging SDKs, without requiring any additional custom instrumentation
from the user.

This OTEP aims at proposing consistent guidelines for creating spans that model
the stages of the messaging flow, and correlating those in a way so that the
requirements sketched above can be met in a consistent way across messaging
scenarios and different messaging systems.

## Terminology

For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).

> ## Stages of producing and consuming messages

Reviewer: I'm totally fine with the repetition generally, my only complaint was that it was unmarked.

measures the duration of the call or operation that provides messages for
sending or publishing to an intermediary. This call or operation (and the
related "Publish" span) can either refer to a single message or to a batch of
multiple messages.

There are three different scenarios for injecting a creation context into a message:

1. A user provides a custom creation context for a message, which doesn't refer
   to any span described in this document. This provides flexibility to users
   to model custom scenarios. In this case, no other spans besides the "Publish"
   span should be created. The "Publish" span should link to the provided
   creation context.
2. If no custom creation context is provided for a message, it is recommended
   to create a "Create" span for every single message and inject its context
   into the message.  "Create" spans can be created during the "Publish" operation
   as children of the "Publish" span.  Alternatively, "Create" spans can be
   created independently of the "Publish" operation, e. g. in cases where messages
   are created before they are passed to a "Publish" operation. In this case, the
   "Publish" span should link to the "Create" spans.

> 3. For single-message scenarios and when large number of spans are a problem,

Reviewer: ```suggestion
3. For single-message scenarios or when large number of spans are a problem,
```

?

   the context of the "Publish" span can be injected into the message, thus
   acting as the creation context. In this case, no other spans besides the
   "Publish" span should be created.

### Consumer

> For many use cases, it is not possible to rely on the presence of "Process"

Reviewer: This seems a bit out of order. Maybe first start with describing what a "Process" span would be (could be very simple, referring to the previous scenario definition). Also, structure-wise it seems odd that "Process" is described directly under "Consumer", even being the only top-level content, while everything else is in subsections.

spans for correlating producer with consumer traces: there are cases where a
dedicated processing operation cannot be identified, or where processing

> happens in a different trace. Furthermore, processing operations are not often

Reviewer: > or where processing happens in a different trace

Deliver and receive will also happen in different traces, due to being only linked and not child spans (impossible in general case with batching).

covered by messaging libraries and SDKs, but take place in application code.
Consistently creating spans for "Processing" operations would require either
effort from the application owner to correctly instrument those operations, or
additional capabilities of messaging libraries and SDKs (e. g. hooks for
processing callbacks, which can then be instrumented by the libraries or SDKs).

While it is possible to create "Process" spans and correlate those with

> consumer traces in certain cases, this is not something that can be generally

Reviewer: Is "producer" meant here? (Cf. previous, maybe "or where processing should happen in the _same_ trace was meant?)

required. Therefore, it is more feasible to require the creation of "Deliver"
spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
correlate producer with consumer traces.

#### Instrumenting push-based scenarios

In push-based consumer scenarios, the delivery of messages is not initiated by
the application code. Instead, callbacks or handlers are registered and then
called by messaging SDKs to forward messages to the application.

A "Deliver" span covers the call of such a callback or handler and should link
to the creation context of all messages that are forwarded via the respective
call.

#### Instrumenting pull-based scenarios

> In pull-based consumer scenarios, the delivery of messages is requested by the
> application code. This usually involves a blocking call, which returns zero or

Reviewer: ```suggestion
In pull-based consumer scenarios, the message is actively requested by the
application code. This usually involves a blocking call, which returns zero or
```

To not use the term "delivery" in the non-delivery case.

Could also use "fetched" instead of "requested".

more messages on completion.

A "Receive" span covers such calls and should link to the creation context of

> all messages that are forwarded via the respective call.

Reviewer: A note could be added here, referring to the Open Question for links after span creation.

to receiving the message from intermediaries, but instead refer to the
application receiving messages for processing. If messages are fetched from the
intermediary and forwarded to the application in one go, the whole operation
might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
might pre-fetch messages from intermediaries and cache those messages, and only
forward messages to the application at a later time. In this case, the
operation of pre-fetching and caching should not be covered by the "Deliver" or
"Receive" spans.

Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
to notify the application that no message is available for processing), one
message, or multiple messages (a batch of messages). "Deliver" and "Receive"
spans should link to the creation context of the messages forwarded, thus those
spans can link to zero, one, or multiple producer spans.

For single-message scenarios, and if the "Deliver" or "Receive" spans would be
root spans of a new trace, the creation context may also be used as a parent on
those operations in addition to being added as a link. Keeping single-messages

> operations in the same trace can greatly improve the user experience.

Reviewer: Can we add: "In any case, instrumentations may also provide a way to explicitly request using a creation context as parent."?

Some backends may need this e.g. if they don't support links well or to improve performance.

Feel free to disregard, as the OTEP is already in a late stage.

  the intermediary.

Settlement operations on the consumer can either be triggered manually by the
user, or can be triggered automatically by messaging SDKs based on return
values of callbacks.

A "Settle" span should be created for every settlement operation that happens
on the consumer (at-least-once and exactly-once). SDKs will, in some cases,
auto-settle messages in push-scenarios when messages are delivered via
callbacks.

"Settle" spans should link to creation context of the messages that are
settled, when possible.

No settlement span should be created for settlement scenarios that do not
involve any settlement operations on the consumer side.

## Proposed changes and additions to the messaging semantic conventions

### Operation name

The following operations related to messages are covered by these semantic
conventions:

| Operation name | Description |
|----------------|-------------|
| `publish`      | One or more messages are provided for publishing to an intermediary. |
| `create`       | A message is created. |
| `receive`      | One or more messages are requested by a consumer. |
| `deliver`      | One or more messages are passed to a consumer. |

> | `settle`       | One or more messages are settled. |

Reviewer: This is a nice summary that could be useful much earlier in the concepts 😉

For further details about each of those operations refer to the [section about trace structure](#trace-structure).

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|

> | `publish`      | `PRODUCER`, if no `create` spans are present. |

Reviewer: Otherwise?

| `create`       | `PRODUCER` |
| `receive`      | `CONSUMER` |

> | `deliver`      | `CONSUMER` |

Reviewer: ```suggestion
| `deliver`      | `CONSUMER` |
|`settle?        | (see below)|
```

(excuse the spacing)

The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
e. g. it should be set to `CLIENT` if the `settle` spans models a synchronous call

> to the intermediary.

Reviewer: Can we just recommend CLIENT? I don't see any other that makes sense in given the actual usages of span kinds in our other (non-messaging) semantic conventions, even if it would fit the abstract span kind definition maybe.

Setting span kinds according to this table ensures that span links between
consumers and producers always exist between a `PRODUCER` span on the producer
side and a `CONSUMER` span on the consumer side. This allows analysis tools to

> interpret linked traces without the need for additional semantic hints.

Reviewer: Not sure what is meant by that, or if this should really be relied upon. Note that there is also https://github.com/open-telemetry/opentelemetry-proto/pull/484 to gain more insights into the trace structure (and the accepted OTEP for it)

> ### Span relationships

Reviewer: This seems like a rehash of the "Trace structure" section? Please unify, or add a note about the relationship of this section to "Trace structure".

Reviewer: Then I propose:

```suggestion
### Span relationships

This section defines semantic conventions for the high-level structure outlined in [Trace structure](#trace-structure) more closely. In case of any contradictions, this section here overrides anything in the former section.
```

Author: I added this paragraph at the beginning of the "Proposed changes" section, I hope this makes things clearer:
 
 > This section contains a concise and normative definition of what was outlined in the [Trace structure](#trace-structure) section. The following subsections are supposed to be merged into the semantic conventions, whereas the detailed description in the [Trace structure](#trace-structure) section in this OTEP will serve as explanation and future reference.

For single-message scenarios, and if the "Deliver" or "Receive" spans would be
root spans of a new trace, the creation context may also be used as a parent on
those operations in addition to being added as a link. Keeping single-messages
operations in the same trace can greatly improve the user experience.

#### Settlement

Messages can be settled in a variety of different ways:

* The intermediary settles the messages as it is sent to the consumer. No
  settlement operations happen on the consumer.
* The consumer settles a message without awaiting an acknowledgment from the
  intermediary.
* The consumer settles a message and awaits an acknowledgment from the
  intermediary. This involves a round-trip exchange between the consumer and
  the intermediary.

Settlement operations on the consumer can either be triggered manually by the
user, or can be triggered automatically by messaging SDKs based on return
values of callbacks.

A "Settle" span should be created for every settlement operation that happens
on the consumer (at-least-once and exactly-once). SDKs will, in some cases,
auto-settle messages in push-scenarios when messages are delivered via

> callbacks. In cases where it is possible, it is recommended to create the

Reviewer 5: ```suggestion
callbacks. 
```

I suggest removing this sentence as it's misleading. If settlement happens in scope of deliver span - great, deliver will be parent automatically. Otherwise, it might be confusing since deliver can end before settlement happens or they may happen in different scopes. Instrumentation authors should not try to fix it. 
I believe we already have an example that demonstrates that it can happen on its own and it should be good enough.

> "Settle" span as a child of the "Deliver" span.

Reviewer 8: I know this is a tired topic in WG discussions, and I think it's much more well-defined now. However, I just want to make sure my last remaining concern here is understood, and that we collectively feel we're making the right trade-off.

My thoughts above are with the following scenario in mind:
* Single message receive
* Only consumer (no intermediary) involved in settlement.
* Auto-ack or client ack from within delivery callback.

I believe this is a very common scenario. It represents one of the simplest messaging scenarios possible (where clients are involved in settlement at all), and why I want to draw attention to it.

The cases where it is possible to create "Settle" as a child of "Deliver" would be in the described use case above.

When the message is settled without waiting for a response from the intermediary (what the AMQP 1.0 specification describes as at least once delivery), I don't think the "end timestamp" provides useful information. The most useful information is when the application performed the settlement. Furthermore, for single message delivery, there's no need to correlate the "settle" with the message creation context since the "Deliver" already does that. IMO this is a very common messaging pattern and is worth considering. There's a trade-off to be made here:
* Create an additional span for the settle operation so that we are consistent and all messages that involve the consumer in settlement always have a settle span. Bear in mind this also means creating a link to the creation context, which is yet more overhead; vs.
* Just create an event in the "Deliver" span with the timestamp the application settled the message. The span already has a link to the creation context.

So ultimately what I'm wondering is whether the value of "every consumer-initiated settlement results in a settle span" worth the cost of providing little other useful information in what I'd consider one of the simplest messaging scenarios where consumers are involved in settlement.

I suppose the normative language would technically allow to place information related to settlement into a delivery span event -- the spec only says that:
> "Settle" spans SHOULD be created for every manually or automatically triggered settlement operation. 

Since SHOULD is used rather than MUST, it does leave it open to omitting the send span and instead including relevant settlement data in a "Deliver" span. But I think an implementation violating a "SHOULD" recommendation would generally be frowned upon, yet I think there's a good justification for omitting the send span in the case I describe above.

Reviewer 5: Let me turn this around a bit and add a third option that would probably work for both points:

- if there is a remote call to settle, it must be a span (to have duration, status, provide scope for transport spans underneath). It comes with a MUST/SHOULD
- if there is no remote call to settle (in memory settlement?), instrumentation (or system-specific messaging convention extension) MAY create a span or MAY instead populate attributes on the deliver span (using certain not-yet-defined attributes to express settlement kind - ack/nack/whatnot)

Author: This was discussed in today's workgroup meeting.

We decided to leave the wording as it is, and have a SHOULD requirement for a settlement span for each settlement operation.

If this doesn't work well for individual cases, they can still omit the settlement span (as it's just a SHOULD requirement).

If this turns out to be a general problem and lots of users are unhappy with it, we can re-discuss and maybe change from SHOULD to MAY. We also want to revisit this topic in the context of settlement attribute discussions.

> callbacks. In cases where it is possible, it is recommended to create the
> "Settle" span as a child of the "Deliver" span.

Reviewer 5: ```suggestion
callbacks. 
```

I suggest removing this sentence as it's misleading. If settlement happens in scope of deliver span - great, deliver will be parent automatically. Otherwise, it might be confusing since deliver can end before settlement happens or they may happen in different scopes. Instrumentation authors should not try to fix it. 
I believe we already have an example that demonstrates that it can happen on its own and it should be good enough.

Reviewer 6: I think it should be fine leaving it out 👍

# Span structure for messaging scenarios

This OTEP aims at defining consistent conventions about what spans to create
for messaging scenarios, and at defining how those spans relate to each other.

This OTEP is based on [OTEP 0173](0173-messaging-semantic-conventions.md),
which defines basic terms and describes messaging scenarios that should be
supported by messaging semantic conventions. It also relies on context
propagation requirements put forth in the existing [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/messaging.md#context-propagation)
and detailed in [OTEP 0205](0205-messaging-semantic-conventions-context-propagation).

* [Terminology](#terminology)
* [Motivation](#motivation)
* [Stages of producing and consuming messages](#stages-of-producing-and-consuming-messages)
* [Trace structure](#trace-structure)

> * [Proposed changes and additions to the messaging semantic conventions](proposed-changes-and-additions-to-the-messaging-semantic-conventions)

Reviewer: Are you missing all the "#"s from here on, or does GH make this work?

Author: Fixed. It seems GH made it work, however, I added the hashes nevertheless.

requirements sketched above can be met in a consistent way across messaging
scenarios and different messaging systems.

## Stages of producing and consuming messages

As previously described in [OTEP 173](https://github.com/open-telemetry/oteps/blob/main/text/trace/0173-messaging-semantic-conventions.md#scenarios.),
producing and consuming a message involves five stages:

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  CR[Create] --> PU[Publish]
  end
  subgraph INTERMEDIARY
  direction TB
  end
  subgraph CONSUMER
  direction TB
  RE[Receive] --> PR[Process]
  PR --> SE[Settle]
  end
  PU --> INTERMEDIARY
  INTERMEDIARY --> RE
  SE -..- INTERMEDIARY
```

1. The producer creates a message.
2. The producer publishes the message to an intermediary.
3. The consumer receives the message from an intermediary.
4. The consumer processes the message.
5. The consumer settles the message by notifying the intermediary that the
   message was processed. In some cases (fire-and-forget scenarios, or when
   settlement happens on the broker), the settlement stage does not exist.

The semantic conventions below define how to model those stages with spans.

> ## Trace structure
> 

Reviewer: ```suggestion
## Trace structure

This section outlines the general structure of messaging traces on a high level. The details are later
described in [Proposed changes and additions to the messaging semantic conventions](#proposed-changes-and-additions-to-the-messaging-semantic-conventions) (note especially the [Span relationships](#span-relationships) subsection).

```

Author: I hope this paragraph in the "Proposed changes" section makes things clearer:

> This section contains a concise and normative definition of what was outlined in the [Trace structure](#trace-structure) section. The following subsections are supposed to be merged into the semantic conventions, whereas the detailed description in the [Trace structure](#trace-structure) section in this OTEP will serve as explanation and future reference.

related "Publish" span) can either refer to a single message or to a batch of
multiple messages.

There are three different scenarios for injecting a creation context into a message:

1. A user provides a custom creation context for a message, which doesn't refer
   to any span described in this document. This provides flexibility to users
   to model custom scenarios. In this case, no other spans besides the "Publish"
   span should be created.
2. If no custom creation context is provided for a message, it is recommended

>    to create a "Create" span for every single message and inject its context.

Reviewer 6: ```suggestion
   to create a "Create" span for every single message and inject its context in the message.
```

Not sure if it's just me, but I had to read it 3x until I got "its context" was the span context in the message.

Author: Applied, just using "into" instead of "to".

   "Create" spans can be created during the "Publish" operation as children of the

>    "Publish" span.  Alternatively, "Create" spans can be created independently of

Reviewer 6: Isn't this sentence:

> Alternatively, "Create" spans can be created independently of the "Publish" operation, e. g. in cases where messages are created before they    are passed to a "Publish" operation

Duplicated/say the same thing as in point 1. above? 

Because "created independently of the "Publish" operation" is essentially "user-land" code and that would fall under the "A user provides a custom creation context for a message" on point 1. I feel we can remove it entirely. 

Also, this part of "Publish" must link to creation is not really related to the "header" of this numbered list (here are three different scenarios for injecting a creation context into a message:). I see the value on talking about the links, so maybe make it a sub-item of 2. instead?

Reviewer 6: We discussed this in the meeting (25.05) and others also agreed on the points above. @dpauls suggested also to remove the sentence "Anternatively..." and add something like this below points 1.

```
1. A user provides a custom creation context for a message, which doesn't refer
   to any span described in this document. This provides flexibility to users
   to model custom scenarios. In this case, no other spans besides the "Publish"
   span should be created, and the "Publish" span should link to this user provided
   context
```

Author: > Because "created independently of the "Publish" operation" is essentially "user-land" code and that would fall under the "A user provides a custom creation context for a message" on point 1. I feel we can remove it entirely.

Messaging SDKs can offer different API calls for creating and for publishing a message, that's the main intention of point (2) in the list above. E. g., this is from a Java Kafka SDK:

```java
final ProducerRecord<K, V> record = new ProducerRecord<>(topic, key, value);
Future<RecordMetadata> future = producer.send(record);
```

The constructor of `ProducerRecord` could be covered by a "Create" span (and its context being attached to the record), `producer.send` could be covered by a "Publish" span. This wouldn't really be "user-land" code, but instrumentation provided by an SDK.

Reviewer 6: I see your point now. I think it's just a matter of how you look at it, since the create span was still created somewhere else and just passed to "send". I don't have any strong opinions on this, so feel free to close if you think what we have now is clear 👍

   the "Publish" operation, e. g. in cases where messages are created before they
   are passed to a "Publish" operation. In this case, the "Publish" span should
   link to the "Create" spans.
3. For single-message scenarios and when large number of spans are a problem,

>    the context of the "Publish" span can be injected as creation context. In

Reviewer 6: ```suggestion
   the context of the "Publish" span can be injected in the message, thus acting as the creation context. In
```

Author: Applied.
